{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": "true"
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Statement-of-Project-Scope\" data-toc-modified-id=\"Statement-of-Project-Scope-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Statement of Project Scope</a></span></li><li><span><a href=\"#General-Introduction\" data-toc-modified-id=\"General-Introduction-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>General Introduction</a></span></li><li><span><a href=\"#Getting-the-Data\" data-toc-modified-id=\"Getting-the-Data-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Getting the Data</a></span><ul class=\"toc-item\"><li><span><a href=\"#SRA-Toolkit-Setup\" data-toc-modified-id=\"SRA-Toolkit-Setup-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>SRA Toolkit Setup</a></span></li><li><span><a href=\"#Downloading-the-First-10,000-Reads\" data-toc-modified-id=\"Downloading-the-First-10,000-Reads-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>Downloading the First 10,000 Reads</a></span></li><li><span><a href=\"#Importing-the-Reads\" data-toc-modified-id=\"Importing-the-Reads-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>Importing the Reads</a></span></li><li><span><a href=\"#About-the-FastQ-File-Format\" data-toc-modified-id=\"About-the-FastQ-File-Format-3.4\"><span class=\"toc-item-num\">3.4&nbsp;&nbsp;</span>About the FastQ File Format</a></span></li></ul></li><li><span><a href=\"#Quality-Control\" data-toc-modified-id=\"Quality-Control-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Quality Control</a></span><ul class=\"toc-item\"><li><span><a href=\"#Visualizing-Read-Statistics\" data-toc-modified-id=\"Visualizing-Read-Statistics-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Visualizing Read Statistics</a></span></li><li><span><a href=\"#Resuming-Trimming\" data-toc-modified-id=\"Resuming-Trimming-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Resuming Trimming</a></span></li><li><span><a href=\"#Running-a-full-pipeline\" data-toc-modified-id=\"Running-a-full-pipeline-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Running a full pipeline</a></span></li><li><span><a href=\"#Aligning-Reads\" data-toc-modified-id=\"Aligning-Reads-4.4\"><span class=\"toc-item-num\">4.4&nbsp;&nbsp;</span>Aligning Reads</a></span></li><li><span><a href=\"#Parsing-BBMap-Output\" data-toc-modified-id=\"Parsing-BBMap-Output-4.5\"><span class=\"toc-item-num\">4.5&nbsp;&nbsp;</span>Parsing BBMap Output</a></span></li><li><span><a href=\"#Confronting-Resource-Limitations\" data-toc-modified-id=\"Confronting-Resource-Limitations-4.6\"><span class=\"toc-item-num\">4.6&nbsp;&nbsp;</span>Confronting Resource Limitations</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Table of Contents**\n",
    "<div id=\"toc\"></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import locale\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import platform\n",
    "import shutil\n",
    "import sys\n",
    "import webbrowser\n",
    "# Necessary to set font family for Latex in matplotlib:\n",
    "from matplotlib import rc\n",
    "\n",
    "locale.setlocale(locale.LC_ALL, '')\n",
    "o_s = platform.system()\n",
    "paths = {}\n",
    "\n",
    "# to make this notebook's output stable across runs\n",
    "np.random.seed(42)\n",
    "\n",
    "if o_s == 'Darwin':\n",
    "    paths['data_dir'] = '/Users/drew/data/Bio'\n",
    "    paths['sra_tools_dir'] = '/Users/drew/Documents/Data/Bio/\\\n",
    "sratoolkit/bin'\n",
    "    paths['fastqc'] = '/Applications/FastQC.app/Contents/MacOS/fastqc'\n",
    "    paths['bbmap_dir'] = '/Users/drew/Documents/Data/Python/bbmap'\n",
    "    \n",
    "elif o_s == 'Windows':\n",
    "    paths['data_dir'] = r'C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap'\n",
    "    paths['sra_tools_dir'] = r'C:\\Users\\DMacKellar\\Documents\\\\\n",
    "Python\\BioPython\\Galaxy_rnaseq\\sratoolkit\\sratoolkit.2.8.2-1-win64\\bin'\n",
    "    paths['fastqc_dir'] = r'C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\FastQC'\n",
    "    paths['multiqc_dir'] = r'C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\MultiQC'\n",
    "    paths['jdk_dir'] = r'C:\\Program Files\\Java\\jdk1.8.0_101'\n",
    "    paths['bzip_dir'] = r'C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\Glimmer\\Glimmer-master\\src\\main\\java'\n",
    "    paths['bbmap_dir'] = r'C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\BBMap'\n",
    "    paths['phi_x'] = r'C:/Users/DMacKellar/Documents/Python/BioPython/BBMap/resources/phix174_ill.ref.fa'\n",
    "    paths['grch38_dir'] = os.path.join(paths['data_dir'], 'grch38')\n",
    "\n",
    "# rc('font',**{'family':'serif','serif':['DejaVu Sans']})\n",
    "rc('text', usetex=False)\n",
    "\n",
    "for root, dirs, files in os.walk(paths['data_dir']):\n",
    "    for file in files:\n",
    "        file_under = str(file).replace('.', '_')\n",
    "        path = os.path.join(root, file)\n",
    "        paths[file_under] = path\n",
    "        \n",
    "sra_table = pd.read_csv(paths['Bmap_SraRunTable_txt'], sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statement of Project Scope\n",
    "\n",
    "The point of this notebook is to brush up on some basic Bioinformatics data-wrangling tasks.  I've sequenced a novel bacterial genome before, but in this case I thought I'd try aligning some RNAseq data, and thereby deriving tissue-specific gene expression patterns.\n",
    "\n",
    "I've selected as an application aligning the commonly-used Illumina BodyMap RNAseq reads to the Human genome.  The raw reads for the Illumina BodyMap 2.0 project are [here](https://www.ncbi.nlm.nih.gov/sra?term=ERP000546&cmd=DetailsSearch).  There are 48 files, each with somewhere around 8E7 spots (potentially representing that many transcripts), for about 6-8E9 Bp per file.  A per-run listing can be found by clicking the above link and following the link at the top that says 'Send results to Run selector'.  I'll use a version of the data derived thereby below, to sort the data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# General Introduction\n",
    "\n",
    "RNA-seq alignment and quantification involves specific assumptions about methodology that are distinct from those of processing DNA reads for genome sequencing.\n",
    "\n",
    "General info about the Illumina Body Map 2 is available from several sources.  One note on methodology from [this talk](https://personal.broadinstitute.org/mgarber/bodymap_schroth.pdf):\n",
    "\n",
    "* The samples used for the 2X50 and 1X75 bp runs are prepared using the Illumina mRNA-Seq kit. \n",
    "    * These libraries are made from poly-A selected mRNA \n",
    "    * They are made with a random priming process and are not stranded (i.e., can represent either F or R direction)\n",
    "    * The insert size for the PE libraries is about ~210 bps\n",
    "        \n",
    "16 Human tissues are represented in the project, and for each tissue they ran one whole lane of a HiSeq2000 sequencer, mixing the paired-end and 1x75 sequences.\n",
    "\n",
    "As a side project, they pooled total RNA from each of the 16 tissues, then subjected the pooled RNAs to each of three different treatments:\n",
    "\n",
    "* Total Poly-A selected mRNA\n",
    "* Total Poly-A selected mRNA with Normalization\n",
    "* Total RNA – no Poly-A Selection – to enrich for non-coding RNA\n",
    "    * Complete Transcriptome Library Prep Method to capture all RNA species\n",
    "    * Uses New Illumina protocol for reducing rRNA in whole transcriptome analysis\n",
    "\n",
    "Apparently, the point of this latter approach was to validate a new protocol they had developed that was meant to enrich non-rRNA in a sample.  From each of these pooled runs they collected 1x100bp reads per sequence.\n",
    "\n",
    "I find that the easiest way to parse the metadata associated with the project is to go to the SRA via the link provided above, go to the bottom of the first table in the right-hand panel, and, under the column 'Downloads', click the button 'RunInfo Table'.  Save the resulting file to a local path and open it with Pandas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSourceProvider</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>InsertSize</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>Library_Name</th>\n",
       "      <th>MBases</th>\n",
       "      <th>MBytes</th>\n",
       "      <th>Run</th>\n",
       "      <th>...</th>\n",
       "      <th>DATASTORE_filetype</th>\n",
       "      <th>DATASTORE_provider</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>SRA_Study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100</td>\n",
       "      <td>SAMEA962337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX011226</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>HCT20170</td>\n",
       "      <td>7290</td>\n",
       "      <td>4098</td>\n",
       "      <td>ERR030856</td>\n",
       "      <td>...</td>\n",
       "      <td>sra</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2014-05-30</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2011-03-17</td>\n",
       "      <td>ERP000546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100</td>\n",
       "      <td>SAMEA962337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX011216</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>HCT20170</td>\n",
       "      <td>7461</td>\n",
       "      <td>4190</td>\n",
       "      <td>ERR030857</td>\n",
       "      <td>...</td>\n",
       "      <td>sra</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2014-05-30</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2011-03-17</td>\n",
       "      <td>ERP000546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100</td>\n",
       "      <td>SAMEA962337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX011224</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>HCT20170</td>\n",
       "      <td>7365</td>\n",
       "      <td>4151</td>\n",
       "      <td>ERR030858</td>\n",
       "      <td>...</td>\n",
       "      <td>sra</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2014-05-30</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2011-03-17</td>\n",
       "      <td>ERP000546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100</td>\n",
       "      <td>SAMEA962346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX011189</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>HCT20172</td>\n",
       "      <td>7274</td>\n",
       "      <td>4097</td>\n",
       "      <td>ERR030859</td>\n",
       "      <td>...</td>\n",
       "      <td>sra</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2014-05-30</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2011-03-17</td>\n",
       "      <td>ERP000546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100</td>\n",
       "      <td>SAMEA962346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX011223</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>HCT20172</td>\n",
       "      <td>7241</td>\n",
       "      <td>4062</td>\n",
       "      <td>ERR030860</td>\n",
       "      <td>...</td>\n",
       "      <td>sra</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2014-05-30</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2011-03-17</td>\n",
       "      <td>ERP000546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   AvgSpotLen    BioSample BioSourceProvider Experiment  InsertSize  \\\n",
       "0         100  SAMEA962337               NaN  ERX011226           0   \n",
       "1         100  SAMEA962337               NaN  ERX011216           0   \n",
       "2         100  SAMEA962337               NaN  ERX011224           0   \n",
       "3         100  SAMEA962346               NaN  ERX011189           0   \n",
       "4         100  SAMEA962346               NaN  ERX011223           0   \n",
       "\n",
       "  LibraryLayout Library_Name  MBases  MBytes        Run    ...      \\\n",
       "0        SINGLE     HCT20170    7290    4098  ERR030856    ...       \n",
       "1        SINGLE     HCT20170    7461    4190  ERR030857    ...       \n",
       "2        SINGLE     HCT20170    7365    4151  ERR030858    ...       \n",
       "3        SINGLE     HCT20172    7274    4097  ERR030859    ...       \n",
       "4        SINGLE     HCT20172    7241    4062  ERR030860    ...       \n",
       "\n",
       "  DATASTORE_filetype DATASTORE_provider           Instrument LibrarySelection  \\\n",
       "0                sra               ncbi  Illumina HiSeq 2000             cDNA   \n",
       "1                sra               ncbi  Illumina HiSeq 2000             cDNA   \n",
       "2                sra               ncbi  Illumina HiSeq 2000             cDNA   \n",
       "3                sra               ncbi  Illumina HiSeq 2000             cDNA   \n",
       "4                sra               ncbi  Illumina HiSeq 2000             cDNA   \n",
       "\n",
       "    LibrarySource    LoadDate      Organism  Platform ReleaseDate  SRA_Study  \n",
       "0  TRANSCRIPTOMIC  2014-05-30  Homo sapiens  ILLUMINA  2011-03-17  ERP000546  \n",
       "1  TRANSCRIPTOMIC  2014-05-30  Homo sapiens  ILLUMINA  2011-03-17  ERP000546  \n",
       "2  TRANSCRIPTOMIC  2014-05-30  Homo sapiens  ILLUMINA  2011-03-17  ERP000546  \n",
       "3  TRANSCRIPTOMIC  2014-05-30  Homo sapiens  ILLUMINA  2011-03-17  ERP000546  \n",
       "4  TRANSCRIPTOMIC  2014-05-30  Homo sapiens  ILLUMINA  2011-03-17  ERP000546  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_table = pd.read_csv(paths['Bmap_SraRunTable_txt'], sep='\\t')\n",
    "\n",
    "sra_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting the Data\n",
    "\n",
    "With the metadata in hand, we can select the reads of interest and prepare to download them programmatically by their accession number.  To interact with the SRA repository, however, NCBI requires that we download a special suite of software that they have written for this purpose; the reads don't appear to be available via any kind of structured URL query.\n",
    "\n",
    "## SRA Toolkit Setup\n",
    "\n",
    "Binaries are available for Linux, Mac OS X, and Windows.  Instructions are [here](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=std).  They can be downloaded and installed on some specified local directory, this location stored in a dict in Python, and called from within this notebook using the `subprocess` package.\n",
    "\n",
    "Selecting the read files that interest you can be done by parsing the `sra_table` that I've downloaded above; the proper accession number for the SRA will be in the `Runs` column.\n",
    "\n",
    "Then, you have to use the NCBI SRA Tools [fastq-dump](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=fastq-dump) program to get the FastQ (or FASTA) file from the raw image data that includes the flow cell the way the sequencer saw it.\n",
    "\n",
    "The command-line syntax of the call to `fastq-dump` is of the general form:\n",
    "\n",
    "    fastq-dump.exe -X 5 -Z SRR390728\n",
    "    \n",
    "in which the arg `-X 5` specifies that you only mean to download the first 5 reads (useful for ensuring the command works for you before you execute it in a large batch; obviously modify or omit this to suit your purposes).\n",
    "\n",
    "Note: the default is to download paired reads from a single sequencing run as a single file, where the pairs are interleaved.  I find it more useful to download them as separate files, in which case the `'split-3'` arg should be passed to `fastq-dump`.\n",
    "\n",
    "Once the general syntax is understood, proceed to download the runs as a batch using the `subprocess` package.  If you're unused to it, [`subprocess`](https://docs.python.org/3.6/library/subprocess.html) comes with significant functionality, but also complications of interface.  It will be useful to check our code by simply printing results below, before executing a full-fledged download:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "if o_s == 'Windows':\n",
    "    exe_ext = '.exe'\n",
    "else:\n",
    "    exe_ext = ''\n",
    "paths['fastq_dump'] = os.path.join(paths['sra_tools_dir'], \n",
    "                                   'fastq-dump'+exe_ext)\n",
    "paths['prefetch'] = os.path.join(paths['sra_tools_dir'], \n",
    "                                 'prefetch'+exe_ext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['/Users/drew/Documents/Data/Bio/sratoolkit/bin/fastq-dump', '-X', '5', '-Z', 'SRR390729'], returncode=0, stdout=b'Read 5 spots for SRR390729\\nWritten 5 spots for SRR390729\\n@SRR390729.1 1 length=100\\nACCCTTCCCAACACCCTGGGAACCTATGGGGCCAGGCGTTCCTTACCAAAGCTCATGAGAAATACATCGAGCACAAAACTCTTCATCAGCTGGTTTTACT\\n+SRR390729.1 1 length=100\\n::::::::::::::::::::::::::::::44444///////////////::::::::::::::::::::::::::1:::44443///////////////\\n@SRR390729.2 2 length=100\\nGGATATTCGTATTCATCTTAGTGGATAAATACCACCTTACTTGGAAAATACTTCATCTGTAAAATAGAGGACTGCATGGTTCTGTGTATTTTAGAGAAGG\\n+SRR390729.2 2 length=100\\n::::::::::::::::::::::::::::::44444/////////--////:::::::::::::0::::::::::::::::44444///////////////\\n@SRR390729.3 3 length=100\\nAAACGCCAACAGAATGGTATTCAGTATGAATGAAAGAACTAAATTTTAACTTTGGTTGACTCATCTTTATAAGATGACCAGGCTAGAGAATCAGAGATCA\\n+SRR390729.3 3 length=100\\n::::::::::::::::::::::::::::::44444//////////////-::::::::::::::::::::::::::::::44443///////////-///\\n@SRR390729.4 4 length=100\\nTAGCATTTCCATGATACAAACTAAATAACTCTATGGATTTTCTAATGGAAACACACACACACATACACACGCAGAAACAGAAGAGAAGTGGGTGAGATAT\\n+SRR390729.4 4 length=100\\n::::::::::::::::::::::::::::::44444/////////////-/::::::::::::::::::::::::::::::44434///////////////\\n@SRR390729.5 5 length=100\\nAATAATTTGCCAAATTTCTCTCCTAAGACCTAATAATGGATAACACAAACCAATAGTCCTTTTGAATAATCATGGCCTAGCTGTGGTTTTAAAAACATAT\\n+SRR390729.5 5 length=100\\n::::::::::::::::::::::::::::::44444///////////////::::::::::::::::::::::::::.:::43343///////////-///\\n')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "os.chdir(paths['data_dir'])\n",
    "\n",
    "commands1 = '{} -X 5 -Z SRR390728'.format(paths['fastq_dump'])\n",
    "commands2 = '{} -X 10MB SRR390728'.format(paths['prefetch'])\n",
    "commands3 = [paths['fastq_dump'], '-X', '5', '-Z', 'SRR390729']\n",
    "commands4 = [paths['fastq_dump'], '-X', '5', '-Z', 'SRR390729']\n",
    "subprocess.run(commands4, stdout=subprocess.PIPE, \n",
    "               stderr=subprocess.STDOUT, shell=False, check=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That looks serviceable.  Now I'll wrap the general syntax into a function with more flexibility:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sra_download(accession_list, *args, **kwargs):\n",
    "    kwargs_new = []\n",
    "    for k, v in kwargs.items():\n",
    "        kwargs_new.append(k)\n",
    "        kwargs_new.append(str(v))\n",
    "    try:\n",
    "        out_dir\n",
    "    except NameError:\n",
    "        pass\n",
    "    else:\n",
    "        if not os.path.isdir(out_dir):\n",
    "            os.mkdir(out_dir)\n",
    "        os.chdir(out_dir)\n",
    "    print('Downloading {} files from the NCBI SRA server...'.format(\n",
    "        len(accession_list)))\n",
    "    for i, acc in enumerate(accession_list):\n",
    "        cmds = [paths['fastq_dump'], *args, *kwargs_new, str(acc)]\n",
    "        p = subprocess.Popen(cmds, stdout=subprocess.PIPE, \n",
    "                             stderr=subprocess.PIPE, shell=False)\n",
    "        out, err = p.communicate()\n",
    "        if len(err) > 0:\n",
    "            print('\\n{:>4} {} returned an error: \\n{}\\n'.format(\n",
    "                i+1, acc, err.decode('ascii')))\n",
    "        else:\n",
    "            print('{:>4} {} downloaded'.format(i+1, acc))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(os.path.join(paths['data_dir'], 'split3'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading 1 files from the NCBI SRA server...\n",
      "   1 SRR390728 downloaded\n"
     ]
    }
   ],
   "source": [
    "args = ['--split-3']\n",
    "kwargs = {'-X': '10000'}\n",
    "acc1 = ['SRR390728']\n",
    "\n",
    "sra_download(acc1, *args, **kwargs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading the First 10,000 Reads\n",
    "\n",
    "Ok, that's working.  Now, for the purposes of trying out additional steps in the pipeline, and keeping everything lightweight-enough to keep in local memory, I'll just download the first 10,000 spots associated with each of the tissues in the BodyMap dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading 48 files from the NCBI SRA server...\n",
      "   1 ERR030858 downloaded\n",
      "   2 ERR030859 downloaded\n",
      "   3 ERR030860 downloaded\n",
      "   4 ERR030861 downloaded\n",
      "   5 ERR030862 downloaded\n",
      "   6 ERR030863 downloaded\n",
      "   7 ERR030864 downloaded\n",
      "   8 ERR030865 downloaded\n",
      "   9 ERR030866 downloaded\n",
      "  10 ERR030867 downloaded\n",
      "  11 ERR030868 downloaded\n",
      "  12 ERR030869 downloaded\n",
      "  13 ERR030870 downloaded\n",
      "  14 ERR030871 downloaded\n",
      "  15 ERR030872 downloaded\n",
      "\n",
      "  16 ERR030873 returned an error: \n",
      "2018-06-16T05:20:31 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:20:31 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4644011486 bytes\n",
      "2018-06-16T05:20:31 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:20:32 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:20:32 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4644011486 bytes\n",
      "2018-06-16T05:20:32 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  17 ERR030874 returned an error: \n",
      "2018-06-16T05:20:41 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:20:41 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4549781976 bytes\n",
      "2018-06-16T05:20:41 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:20:42 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:20:42 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4549781976 bytes\n",
      "2018-06-16T05:20:42 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  18 ERR030875 returned an error: \n",
      "2018-06-16T05:20:50 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:20:50 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4591110380 bytes\n",
      "2018-06-16T05:20:50 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:20:52 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:20:52 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4591110380 bytes\n",
      "2018-06-16T05:20:52 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  19 ERR030876 returned an error: \n",
      "2018-06-16T05:21:00 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:00 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4503544683 bytes\n",
      "2018-06-16T05:21:00 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:21:01 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:01 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4503544683 bytes\n",
      "2018-06-16T05:21:01 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  20 ERR030877 returned an error: \n",
      "2018-06-16T05:21:10 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:10 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4274557907 bytes\n",
      "2018-06-16T05:21:10 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:21:11 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:11 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4274557907 bytes\n",
      "2018-06-16T05:21:11 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  21 ERR030878 returned an error: \n",
      "2018-06-16T05:21:19 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:19 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4034661681 bytes\n",
      "2018-06-16T05:21:19 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:21:21 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:21 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4034661681 bytes\n",
      "2018-06-16T05:21:21 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  22 ERR030879 returned an error: \n",
      "2018-06-16T05:21:30 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:30 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4183525678 bytes\n",
      "2018-06-16T05:21:30 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:21:32 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:32 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4183525678 bytes\n",
      "2018-06-16T05:21:32 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  23 ERR030880 returned an error: \n",
      "2018-06-16T05:21:40 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:40 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4086763666 bytes\n",
      "2018-06-16T05:21:40 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:21:41 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:41 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4086763666 bytes\n",
      "2018-06-16T05:21:41 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  24 ERR030881 returned an error: \n",
      "2018-06-16T05:21:51 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:51 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 3663840382 bytes\n",
      "2018-06-16T05:21:51 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:21:53 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:21:53 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 3663840382 bytes\n",
      "2018-06-16T05:21:53 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  25 ERR030882 returned an error: \n",
      "2018-06-16T05:22:17 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:17 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4211553402 bytes\n",
      "2018-06-16T05:22:17 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:22:19 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:19 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4211553402 bytes\n",
      "2018-06-16T05:22:19 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  26 ERR030883 returned an error: \n",
      "2018-06-16T05:22:27 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:27 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 3747076106 bytes\n",
      "2018-06-16T05:22:27 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:22:29 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:29 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 3747076106 bytes\n",
      "2018-06-16T05:22:29 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  27 ERR030884 returned an error: \n",
      "2018-06-16T05:22:37 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:37 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4465828573 bytes\n",
      "2018-06-16T05:22:37 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:22:38 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:38 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4465828573 bytes\n",
      "2018-06-16T05:22:38 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  28 ERR030885 returned an error: \n",
      "2018-06-16T05:22:46 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:46 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4329482848 bytes\n",
      "2018-06-16T05:22:46 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:22:48 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:48 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4329482848 bytes\n",
      "2018-06-16T05:22:48 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  29 ERR030886 returned an error: \n",
      "2018-06-16T05:22:56 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:56 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4251850780 bytes\n",
      "2018-06-16T05:22:56 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:22:57 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:22:57 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4251850780 bytes\n",
      "2018-06-16T05:22:57 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  30 ERR030887 returned an error: \n",
      "2018-06-16T05:23:07 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:07 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4219250662 bytes\n",
      "2018-06-16T05:23:07 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:23:08 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:08 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4219250662 bytes\n",
      "2018-06-16T05:23:08 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  31 ERR030888 returned an error: \n",
      "2018-06-16T05:23:17 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:17 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4269678434 bytes\n",
      "2018-06-16T05:23:17 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:23:18 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:18 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4269678434 bytes\n",
      "2018-06-16T05:23:18 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  32 ERR030889 returned an error: \n",
      "2018-06-16T05:23:26 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:26 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4214590189 bytes\n",
      "2018-06-16T05:23:26 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:23:28 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:28 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4214590189 bytes\n",
      "2018-06-16T05:23:28 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  33 ERR030857 returned an error: \n",
      "2018-06-16T05:23:36 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:36 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4393870053 bytes\n",
      "2018-06-16T05:23:36 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:23:38 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:38 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4393870053 bytes\n",
      "2018-06-16T05:23:38 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  34 ERR030856 returned an error: \n",
      "2018-06-16T05:23:47 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:47 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4297829576 bytes\n",
      "2018-06-16T05:23:47 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:23:48 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:23:48 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4297829576 bytes\n",
      "2018-06-16T05:23:48 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "  35 ERR030890 downloaded\n",
      "\n",
      "  36 ERR030891 returned an error: \n",
      "2018-06-16T05:26:47 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:26:47 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4304747593 bytes\n",
      "2018-06-16T05:26:47 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:26:48 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:26:48 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4304747593 bytes\n",
      "2018-06-16T05:26:48 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  37 ERR030892 returned an error: \n",
      "2018-06-16T05:26:58 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:26:58 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4514127484 bytes\n",
      "2018-06-16T05:26:58 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:26:59 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:26:59 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4514127484 bytes\n",
      "2018-06-16T05:26:59 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  38 ERR030893 returned an error: \n",
      "2018-06-16T05:27:08 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:08 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4451723433 bytes\n",
      "2018-06-16T05:27:08 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:27:10 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:10 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4451723433 bytes\n",
      "2018-06-16T05:27:10 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  39 ERR030894 returned an error: \n",
      "2018-06-16T05:27:28 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:28 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4384862345 bytes\n",
      "2018-06-16T05:27:28 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:27:30 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:30 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4384862345 bytes\n",
      "2018-06-16T05:27:30 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  40 ERR030895 returned an error: \n",
      "2018-06-16T05:27:38 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:38 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4306556654 bytes\n",
      "2018-06-16T05:27:38 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:27:40 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:40 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4306556654 bytes\n",
      "2018-06-16T05:27:40 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  41 ERR030896 returned an error: \n",
      "2018-06-16T05:27:48 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:48 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4389038264 bytes\n",
      "2018-06-16T05:27:48 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:27:49 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:49 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4389038264 bytes\n",
      "2018-06-16T05:27:49 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  42 ERR030897 returned an error: \n",
      "2018-06-16T05:27:58 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:58 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4474485461 bytes\n",
      "2018-06-16T05:27:58 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:27:59 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:27:59 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4474485461 bytes\n",
      "2018-06-16T05:27:59 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  43 ERR030898 returned an error: \n",
      "2018-06-16T05:28:15 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:15 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4554494573 bytes\n",
      "2018-06-16T05:28:15 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:28:16 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:16 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4554494573 bytes\n",
      "2018-06-16T05:28:16 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  44 ERR030899 returned an error: \n",
      "2018-06-16T05:28:25 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:25 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4522881547 bytes\n",
      "2018-06-16T05:28:25 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:28:26 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:26 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4522881547 bytes\n",
      "2018-06-16T05:28:26 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  45 ERR030900 returned an error: \n",
      "2018-06-16T05:28:34 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:34 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4564540209 bytes\n",
      "2018-06-16T05:28:34 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:28:36 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:36 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4564540209 bytes\n",
      "2018-06-16T05:28:36 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  46 ERR030901 returned an error: \n",
      "2018-06-16T05:28:45 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:45 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4379516773 bytes\n",
      "2018-06-16T05:28:45 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:28:46 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:46 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4379516773 bytes\n",
      "2018-06-16T05:28:46 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  47 ERR030902 returned an error: \n",
      "2018-06-16T05:28:55 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:55 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4539800668 bytes\n",
      "2018-06-16T05:28:55 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:28:56 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:28:56 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4539800668 bytes\n",
      "2018-06-16T05:28:56 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "\n",
      "  48 ERR030903 returned an error: \n",
      "2018-06-16T05:29:04 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:29:04 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4346571367 bytes\n",
      "2018-06-16T05:29:04 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "2018-06-16T05:29:06 fastq-dump.2.8.2 err: unknown while updating file within file system module - unknown system error 'No space left on device(28)'\n",
      "2018-06-16T05:29:06 fastq-dump.2.8.2 int: unknown while updating file within file system module - cannot size local file to 4346571367 bytes\n",
      "2018-06-16T05:29:06 fastq-dump.2.8.2 int: no error - skipping the cache-tee completely\n",
      "\n",
      "\n",
      "CPU times: user 163 ms, sys: 410 ms, total: 574 ms\n",
      "Wall time: 9min 19s\n"
     ]
    }
   ],
   "source": [
    "# %%time\n",
    "\n",
    "# # DCM Note: After running this once locally, comment out\n",
    "# # so we don't repeat the download whenever notebook is run\n",
    "\n",
    "# args = ['--split-3']\n",
    "# kwargs = {'-X': '10000'}\n",
    "# accession_list = sra_table['Run']\n",
    "\n",
    "# sra_download(accession_list, *args, **kwargs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing the Reads\n",
    "\n",
    "Now that I've downloaded a sample of the data sets, let's split them into the relevant experimental condition: the single-end 75bp reads, the paired 50bp reads, and the single-end 100bp reads (this last, remember, comes from pooling the RNA from the various tissues, and is probably uninteresting for this analysis).  I'll write a function below to automatically name the variables that will hold the reads in local memory, automatically appending the relevant condition to the name based on the naming convention with which the reads were deposited in the SRA.  This will obviously vary based on the experiment, but I think it worthwhile to translate the arbitrary accession numbers into tissue-specific monikers when interpreting the data.\n",
    "\n",
    "I'll import the tissue-specific paired- and single-end reads with [Biopython's SeqIO module](https://biopython.org/wiki/SeqIO)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def bmap_reads_import(path=None):\n",
    "    exps, reads = {}, {}\n",
    "    bad = ['16 Tissues mixture', np.NaN]\n",
    "    sra_tiss_df = sra_table[~sra_table.loc[:, 'organism_part'].isin(bad)]\n",
    "    sra_tiss_df_s = sra_tiss_df.sort_values(by=['organism_part', 'LibraryLayout'])\n",
    "    conv_dict = {'_s': '', '_f': '_1', '_r': '_2'}\n",
    "    for _, row in sra_tiss_df_s.iterrows():\n",
    "        if row['LibraryLayout'] == 'SINGLE':\n",
    "            key = row['organism_part']+'_s'\n",
    "            exps[key] = row['Run']\n",
    "        else:\n",
    "            exps[row['organism_part']+'_f'] = row['Run']\n",
    "            exps[row['organism_part']+'_r'] = row['Run']\n",
    "    for exp, run in exps.items():\n",
    "        if exp not in reads.keys():\n",
    "            f_name = run+conv_dict[exp[-2:]]+'.fastq'\n",
    "            f_path = os.path.join(path, f_name)\n",
    "            if os.path.isfile(f_path):\n",
    "                with open(f_path, 'r') as f:\n",
    "                    reads[exp] = SeqIO.to_dict(SeqIO.parse(f, 'fastq'))\n",
    "            \n",
    "    return reads\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 15.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "paths['split3'] = os.path.join(paths['data_dir'], 'split3')\n",
    "bmap_reads = bmap_reads_import(path=paths['split3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('CTGCTTGCAACTANAGCAACAGCCTTCATAGGCTATGNNCTCCCGAGAGG', SingleLetterAlphabet()), id='ERR030880.1', name='ERR030880.1', description='ERR030880.1 HWI-BRUNOP16X_0001:1:1:3016:1084#0 length=50', dbxrefs=[])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's just check one of the records\n",
    "first_rec = bmap_reads['adipose_f']['ERR030880.1']\n",
    "first_rec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSourceProvider</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>InsertSize</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>Library_Name</th>\n",
       "      <th>MBases</th>\n",
       "      <th>MBytes</th>\n",
       "      <th>Run</th>\n",
       "      <th>...</th>\n",
       "      <th>DATASTORE_filetype</th>\n",
       "      <th>DATASTORE_provider</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>SRA_Study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>100</td>\n",
       "      <td>SAMEA962342</td>\n",
       "      <td>Human adrenal total RNA, lot 0812003</td>\n",
       "      <td>ERX011217</td>\n",
       "      <td>248</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>HCT20159</td>\n",
       "      <td>7102</td>\n",
       "      <td>3494</td>\n",
       "      <td>ERR030881</td>\n",
       "      <td>...</td>\n",
       "      <td>sra</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2014-05-30</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2011-03-17</td>\n",
       "      <td>ERP000546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>75</td>\n",
       "      <td>SAMEA962342</td>\n",
       "      <td>Human adrenal total RNA, lot 0812003</td>\n",
       "      <td>ERX011198</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>HCT20159</td>\n",
       "      <td>5448</td>\n",
       "      <td>4019</td>\n",
       "      <td>ERR030889</td>\n",
       "      <td>...</td>\n",
       "      <td>sra</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2014-05-30</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2011-03-17</td>\n",
       "      <td>ERP000546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AvgSpotLen    BioSample                     BioSourceProvider Experiment  \\\n",
       "25         100  SAMEA962342  Human adrenal total RNA, lot 0812003  ERX011217   \n",
       "33          75  SAMEA962342  Human adrenal total RNA, lot 0812003  ERX011198   \n",
       "\n",
       "    InsertSize LibraryLayout Library_Name  MBases  MBytes        Run  \\\n",
       "25         248        PAIRED     HCT20159    7102    3494  ERR030881   \n",
       "33           0        SINGLE     HCT20159    5448    4019  ERR030889   \n",
       "\n",
       "      ...     DATASTORE_filetype DATASTORE_provider           Instrument  \\\n",
       "25    ...                    sra               ncbi  Illumina HiSeq 2000   \n",
       "33    ...                    sra               ncbi  Illumina HiSeq 2000   \n",
       "\n",
       "   LibrarySelection   LibrarySource    LoadDate      Organism  Platform  \\\n",
       "25             cDNA  TRANSCRIPTOMIC  2014-05-30  Homo sapiens  ILLUMINA   \n",
       "33             cDNA  TRANSCRIPTOMIC  2014-05-30  Homo sapiens  ILLUMINA   \n",
       "\n",
       "   ReleaseDate  SRA_Study  \n",
       "25  2011-03-17  ERP000546  \n",
       "33  2011-03-17  ERP000546  \n",
       "\n",
       "[2 rows x 30 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_table[sra_table['organism_part'] == 'adrenal']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## About the FastQ File Format\n",
    "\n",
    "It should be understood that the output above, by which BioPython parses the various data within the FastQ file, contains more than just the unique sequence name and the sequence data that are held in a FASTA file.  Most important is probably the 'quality' of each read, essentially the confidence of the sequencing machine & algorithm about the identity of that nucleotide at the time of sequencing.  But we'll review all of the parts of the file in its native format:\n",
    "\n",
    "In the case of the example files downloaded from the site listed above, the first three reads in '`ERR030881_1.fastq`' are:\n",
    "\n",
    "    @ERR030881.107 HWI-BRUNOP16X_0001:2:1:13663:1096#0/1\n",
    "    ATCTTTTGTGGCTACAGTAAGTTCAATCTGAAGTCAAAACCAACCAATTT\n",
    "    +\n",
    "    5.544,444344555CC?CAEF@EEFFFFFFFFFFFFFFFFFEFFFEFFF\n",
    "    @ERR030881.311 HWI-BRUNOP16X_0001:2:1:18330:1130#0/1\n",
    "    TCCATACATAGGCCTCGGGGTGGGGGAGTCAGAAGCCCCCAGACCCTGTG\n",
    "    +\n",
    "    GFFFGFFBFCHHHHHHHHHHIHEEE@@@=GHGHHHHHHHHHHHHHHHHHH\n",
    "    @ERR030881.1487 HWI-BRUNOP16X_0001:2:1:4144:1420#0/1\n",
    "    GTATAACGCTAGACACAGCGGAGCTCGGGATTGGCTAAACTCCCATAGTA\n",
    "    +\n",
    "    55*'+&&5'55('''888:8FFFFFFFFFF4/1;/4./++FFFFF=5:E#\n",
    "\n",
    "As can be seen, the reads are all cheek-to-jowl, with no lines separating them.  The eye is drawn to the whitespace of the lines that contain only '`+`', but that line doesn't come between different reads, but rather separates the lines containing base pair calls and their respective quality scores.  In fact, the line order goes:\n",
    "\n",
    "* Unique Read Identifier; details on interpretation [available from Wikipedia](https://en.wikipedia.org/wiki/FASTQ_format#Illumina_sequence_identifiers).  The first few characters that are in common among all reads in the file are usually either designate the specific machine on which they were sequenced or, as here, are the string under which the data were registered on the [NCBI Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra).  Specifically, in this case, the page is [here](https://www.ncbi.nlm.nih.gov/sra/?term=ERR030881).\n",
    "\n",
    "* Base Calls; I believe the only valid values are ['A', 'C', 'G', 'T', and 'N'].\n",
    "\n",
    "* A spacer line containing '+'.  Despite appearance, it doesn't necessarily denote any kind of assumption about which strand of DNA in the chromosome from which the read originates; they're all supposed to be '+'.\n",
    "\n",
    "* Quality scores; base quality scores can be interpreted with the info [here](http://www.drive5.com/usearch/manual/quality_score.html).  The encoding apparently differs between platform, however, so attention must be paid to what machine/technology generated the reads.\n",
    "\n",
    "After those four lines, the next read comes immediately, with no separating line.\n",
    "\n",
    "---\n",
    "\n",
    "# Quality Control\n",
    "\n",
    "It will be noted that the extremely powerful and cheap approach of Illumina's 2nd-gen tech to generate sequence data can output results whose quality varies greatly on the skill of the experimenter to generate the data.  The result is that you absolutely can't just dump all of the reads into a mapping protocol and expect much out.  Rather, as indicated above, individual bases within individual reads retain info about how believable they look, and we'll need to first summarize, and then delete some of the data, accordingly, to get this pipeline to run.\n",
    "\n",
    "A very popular (and old) approach to quickly summarizing the quality data associated with any large FASTQ file of sequencing reads is [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/).  I have tried writing a (pretty complicated) custom function in native Python to parse and visualize the output of the reads from BioPython, and it was a little more customized and pretty than FastQC, but it failed to deal with the variable-length reads output after trimming them; I think that, in this and many other steps of this pipeline, wrapping a fully-supported package that someone else has prepared is a more sensible approach.\n",
    "\n",
    "FastQC is generally meant to be run in a GUI mode, outputting `.html` files that can be easily scanned through.  This will result in a workflow that automatically reads all sequences within and offers several tabs of summary data, and prints an icon beside the title of each tab that summarizes whether or not FastQC considers the output problematic for future integration into a mapping/assembly pipeline.  For a more programmatic interface, you have to dig a bit to find the proper syntax to call it from the command line.\n",
    "\n",
    "Furthermore, even once FastQC has run, if you are summarizing all of the datasets in the BodyMap data, you'll have too many outputs to check.  While searching for a good wrapper/analysis tool for the FastQC data, I came upon another program called [MultiQC](http://multiqc.info/docs/).  It looks like it should be suitable for my purposes.  It runs fine on Mac and Linux, but in the process of attempting to get it to work on my PC, I ran into issues that led me to contact the developer.  It was interesting to run down the bugs, but I'll skip any comment on its operation here and get straight to showing the results of the QC.\n",
    "\n",
    "I had wanted to be able to open the output of the QC steps within this notebook, but found that the formatting of the MultiQC tool's output HTML conflicts with native Jupyter functions, essentially breaking the notebook.  So I wrote the function below to open it in a separate tab."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualizing Read Statistics\n",
    "\n",
    "Ok, now it's time to check the reads' lengths, aggregate quality scores, and various other attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_fastqc(trim_dir, *args, **kwargs):\n",
    "    os.chdir(trim_dir)\n",
    "    kwargs_new = []\n",
    "    for k, v in kwargs.items():\n",
    "        kwargs_new.append('{}={}'.format(k, v))\n",
    "    files = []\n",
    "    for file in os.listdir(trim_dir):\n",
    "        path = os.path.abspath(file)\n",
    "        _, extension = os.path.splitext(path)\n",
    "        if extension in ['.fq', '.fastq']:\n",
    "            files.append(path)\n",
    "    if o_s == 'Darwin':\n",
    "        fastqc = paths['fastqc']\n",
    "        multiqc = 'multiqc .'\n",
    "    if o_s == 'Windows':\n",
    "        fastqc = 'java -cp {};{} \\\n",
    "uk.ac.babraham.FastQC.FastQCApplication'.format(paths['bzip_dir'], paths['fastqc_dir'])\n",
    "        multiqc = 'python {} .'.format(os.path.join(*[paths['multiqc_dir'], 'scripts', 'multiqc']))\n",
    "    cmds = [fastqc, *args, *kwargs_new, *files]\n",
    "    final_cmds = ' '.join(cmds)\n",
    "\n",
    "    p = subprocess.Popen(final_cmds, stdout=subprocess.PIPE, \n",
    "                         stderr=subprocess.PIPE, shell=True)\n",
    "    out, err = p.communicate()\n",
    "    if p.returncode != 0:\n",
    "        print('\\n returned an error: \\n{}\\n'.format(\n",
    "            err.decode('ascii')))\n",
    "    else:\n",
    "        print('FastQC run')\n",
    "    p = subprocess.Popen(multiqc, stdout=subprocess.PIPE, \n",
    "                         stderr=subprocess.PIPE, shell=True)\n",
    "    out, err = p.communicate()\n",
    "    if p.returncode != 0:\n",
    "        print('\\n returned an error: \\n{}\\n'.format(\n",
    "            err.decode('ascii')))\n",
    "    else:\n",
    "        print('MultiQC run')\n",
    "        report = os.path.join(trim_dir, 'multiqc_report.html')\n",
    "        webbrowser.open(report)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC run\n",
      "MultiQC run\n"
     ]
    }
   ],
   "source": [
    "run_fastqc(trim_dirs['1'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Resuming Trimming\n",
    "\n",
    "Now, to re-run [bbduk](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide/) with different trimming options.  The first file actually has pretty good quality per base scores out to about 30bp; but the second file (`ERR030872_2_t`) has mean scores below 20 up until base 11.  Further, the mean quality-per-read scores are unacceptably low for a large number of reads in both files.  This might not be too surprising, since the instructions I used for the first run were taken from a section entitled 'adapter trimming', and it didn't do much.  I think I'll need to perform different iterations in series, starting with getting a better trimming run.\n",
    "\n",
    "It appears that it's not recognizing the adapters in the reference file.  The bbduk docs say that they include a more comprehensive list of Illumina adapters in their distro under `/bbmap/resources/adapters.fa`.  I'll point to that and try again.\n",
    "\n",
    "After trying a few different configurations, it looks like a good mix for adapter trimming is to reduce the value of the integers passed to the `k` and `mink` parameters.  The bbduk docs had said that the value of `k` should 'be at most the length of the adapters', which I took to mean larger was better, but it's the kmer used to search the adapter sequences against your seq, so specifying a very long match might be too stringent.  Values of `k` and `mink` around `12` and `3`, respectively, seem to reduce most of the overrepresentation of ACTG in the first dozen bases of both reads.\n",
    "\n",
    "Using the adapters file provided with bbmap causes more reads to be removed than with the custom file I made (with k=12 & mink=3, the % of bases removed is 21% as opposed to 7.5%, respectively.  Alternatively, increasing the hamming distance from `1` to `2` improves 5' variability, trimming ~16% of base info."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting trimming run with args:\n",
      "('tbo -Xmx4g',)\n",
      "['ref=C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\illum_seq.fasta', 'ktrim=l', 'k=12', 'mink=3', 'hdist=2']\n",
      "...\n",
      "\n",
      "   1 ['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\ERR030872_1.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\ERR030872_2.fastq'] returned an error: \n",
      "Executing jgi.BBDukF [tbo, -Xmx4g, ref=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\illum_seq.fasta, ktrim=l, k=12, mink=3, hdist=2, in1=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\ERR030872_1.fastq, in2=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\ERR030872_2.fastq, out1=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\trim_2018-07-08T18_48_05\\ERR030872_1_t.fastq, out2=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\trim_2018-07-08T18_48_05\\ERR030872_2_t.fastq]\n",
      "Version 38.07\n",
      "\n",
      "maskMiddle was disabled because useShortKmers=true\n",
      "0.073 seconds.\n",
      "Initial:\n",
      "Memory: max=1836m, total=124m, free=108m, used=16m\n",
      "\n",
      "Added 46120 kmers; time: \t0.285 seconds.\n",
      "Memory: max=1836m, total=124m, free=102m, used=22m\n",
      "\n",
      "Input is being processed as paired\n",
      "Started output streams:\t0.028 seconds.\n",
      "Processing time:   \t\t0.345 seconds.\n",
      "\n",
      "Input:                  \t20000 reads \t\t1000000 bases.\n",
      "KTrimmed:               \t19951 reads (99.76%) \t166412 bases (16.64%)\n",
      "Trimmed by overlap:     \t4 reads (0.02%) \t9 bases (0.00%)\n",
      "Total Removed:          \t778 reads (3.89%) \t166421 bases (16.64%)\n",
      "Result:                 \t19222 reads (96.11%) \t833579 bases (83.36%)\n",
      "\n",
      "Time:                         \t0.669 seconds.\n",
      "Reads Processed:       20000 \t29.91k reads/sec\n",
      "Bases Processed:       1000k \t1.50m bases/sec\n",
      "\n",
      "\n",
      "FastQC run\n",
      "MultiQC run\n"
     ]
    }
   ],
   "source": [
    "paths['bbmap_adapt'] = os.path.join(*[paths['bbmap_dir'], 'resources', 'adapters.fa'])\n",
    "\n",
    "os.chdir(paths['split3'])\n",
    "kwargs1 = {'ref': paths['bbmap_adapt'],\n",
    "#           'ref': paths['illum_seq_fasta'],\n",
    "          'ktrim': 'l', 'k': 15, 'mink': 3,\n",
    "          'hdist': 1}\n",
    "kwargs2 = {'ref': paths['illum_seq_fasta'],\n",
    "          'ktrim': 'l', 'k': 12, 'mink': 3,\n",
    "          'hdist': 2}\n",
    "args = ['tbo -Xmx4g']\n",
    "\n",
    "# run_bbduk([thyroid_paired], *args, **kwargs1)\n",
    "trim_dirs = run_bbduk([thyroid_paired], *args, **kwargs2)\n",
    "\n",
    "run_fastqc(trim_dirs['1'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running a full pipeline\n",
    "\n",
    "After reviewing the [bbduk docs](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide/) (also elaborated upon [here](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/data-preprocessing/)), I see that the author(s) actually recommend a specific, explicit workflow:\n",
    "\n",
    "* Adapter Trimming\n",
    "* Quality Trimming\n",
    "* Force Trimming\n",
    "* Force-Trim Modulo\n",
    "* Quality Filtering\n",
    "* Kmer Filtering\n",
    "\n",
    "And additional steps that I don't quite understand, but sound like they're of diminishing importance.  I'll try running the 7 steps above in succession using the default args and kwargs listed in the docs, and check the result.\n",
    "\n",
    "---\n",
    "\n",
    "Ok, after having run the entire pipeline with the default params, I see that it's too aggressive.  It reduces the input 10,000 reads down to about 5,500, most of which (~3,000) are 40bp for the forward file, and 35bp for the reverse file, but each file also has significant size spikes at every multiple of 5bp in length (30, 25, etc.).  Per sequence quality scores and mean scores per base are both high; the data that remain would perform very well, I think.  But I'd like to reduce the amount of loss, I think some of the reads/bases being discarded could have value.\n",
    "\n",
    "According to the printed output below, it seems that the biggest losses are occurring at the steps of Quality trimming (ditches 32% of reads and 45% of the bases) and Force trimming (ditches 17% of reads and 31% of bases).  Losses at other steps are marginal.  Actually, the Quality trimming loss was partially because I hadn't used the default quality cutoff of 10 from the right side, but 15 from either side.  Also, with adapter trimming at the default values it's not really catching anything, which could be part of the problem with quality at later steps.  I'll return that step to the params above.\n",
    "\n",
    "Actually, it looks like the first few bases are still too problematic, not in terms of quality scores necessarily, but in terms of variability in distribution of ATCG values.  It's a problem for the first 10 in the first file, but only the first 5 in the second.  I think I'll return the force trim left to 10, and add a length filtering step of 20 at the penultimate step.\n",
    "\n",
    "This approach keeps only ~4800 reads per file, with lengths between 20 and 40bp (mostly on the higher side), but the output quality scores are excellent.  I may provide another variant of this workflow that's less aggressive, but this does look kind of like the best way to recover useable data from the thyroid reads.  I'll wrap this workflow into a function and scale up with more reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note: pass in fastq files as a list of lists; pairs together\n",
    "def run_bbduk_pipeline(fastq_list, trim_dirs=None, use_defaults=True, \n",
    "                       args_dict=None, kwargs_dict=None, \n",
    "                       keep_intermediate_files=False):\n",
    "    def return_default_dicts():\n",
    "        default_kwargs = {'adapter_trimming': {'ref':paths['illum_seq_fasta'], \n",
    "                                           'ktrim':'l', 'k':'15', \n",
    "                                           'mink':'4', 'hdist':'1'},\n",
    "                          'quality_trimming': {'qtrim':'rl', 'trimq':'10'},\n",
    "                          'force_trimming': {'ftl':'10', 'ftr':'139'},\n",
    "                          'force-trim_modulo':{'ftm':'5'},\n",
    "                          'quality_filtering':{'maq':'10', 'minlen':'20'},\n",
    "                          'kmer_filtering':{'ref':paths['phi_x'], 'k':'31', \n",
    "                                            'hdist':'1'}\n",
    "                         }\n",
    "        default_args = {}\n",
    "        for k in default_kwargs.keys():\n",
    "            default_args[k] = []\n",
    "        return default_args, default_kwargs\n",
    "    \n",
    "    if trim_dirs==None:\n",
    "        trim_dirs = {}\n",
    "        new_dir = '1'\n",
    "    else:\n",
    "        new_dir = max([int(x) for x in trim_dirs.keys()]) + 1\n",
    "    if o_s == 'Windows':\n",
    "        bbduk = 'java -cp {}\\current jgi.BBDukF'.format(paths['bbmap_dir'])\n",
    "    elif o_s == 'Darwin':\n",
    "        bbduk = paths['bbduk']\n",
    "    else:\n",
    "        print('Error setting operating system')\n",
    "    try:\n",
    "        out_dir\n",
    "    except NameError:\n",
    "        ts = datetime.datetime.now().replace(microsecond=0).timestamp()\n",
    "        read_ts = datetime.datetime.fromtimestamp(ts).isoformat().replace(':', '_')\n",
    "        out_dir = os.path.join(os.getcwd(), 'trim_{}'.format(read_ts))\n",
    "        os.mkdir(out_dir)\n",
    "    else:\n",
    "        if not os.path.isdir(out_dir):\n",
    "            os.mkdir(out_dir)\n",
    "    trim_dirs[str(new_dir)] = str(out_dir)\n",
    "    os.chdir(out_dir)\n",
    "    \n",
    "    if use_defaults:\n",
    "        args, kwargs = return_default_dicts()\n",
    "    if args_dict:\n",
    "        for k, v in args_dict.items():\n",
    "            for k2, v2 in v.items():\n",
    "                args[k][k2] = v2\n",
    "    if kwargs_dict:\n",
    "        for k, v in kwargs_dict.items():\n",
    "            for k2, v2 in v.items():\n",
    "                kwargs[k][k2] = v2\n",
    "    \n",
    "    for k, v in kwargs.items():\n",
    "        for k2, v2 in v.items():\n",
    "            args[k].append('{}={}'.format(k2, v2))\n",
    "    \n",
    "    with open('trim_log.txt', 'w') as f:\n",
    "        for k, v in args.items():\n",
    "            f.write('{}: {}\\n'.format(k, ' '.join(v)))\n",
    "    print('Starting trimming run with args:\\n')\n",
    "    for k, v in args.items():\n",
    "        print('{}: {}'.format(k, ' '.join(v)))\n",
    "    print('\\n\\nRead files processed:')\n",
    "    def get_next_step(arg_list):\n",
    "        order_of_steps = {'format_conversion': 0, 'adapter_trimming': 1, \n",
    "                          'contaminant_filtering': 2, 'nextera_lmp_lib_split': 3,\n",
    "                          'human_contam_removal': 4, 'quality_trimming': 5, \n",
    "                          'force_trimming': 6, 'force-trim_modulo': 7,\n",
    "                          'quality_filtering': 8, 'kmer_filtering': 9,\n",
    "                          'kmer_masking': 10, 'quality_recalib': 11, \n",
    "                          'deduplication': 12, 'normalization': 13}\n",
    "        sorted_steps = sorted(arg_list, key=lambda x: order_of_steps[x])\n",
    "        for step in sorted_steps:\n",
    "            yield step\n",
    "        \n",
    "    for i, fq in enumerate(fastq_list):\n",
    "        steps_run = []\n",
    "        steps = get_next_step(args)\n",
    "        if len(fq) > 1:\n",
    "            fq_base = [os.path.splitext(os.path.basename(x))[0] for x in fq]\n",
    "        else:\n",
    "            fq_base = [os.path.splitext(os.path.basename(*fq))[0]]\n",
    "        filepaths = [os.path.join(out_dir, x) for x in fq_base]\n",
    "        for step in steps:\n",
    "            if len(steps_run) == 0:\n",
    "                in_suff = ''\n",
    "                out_suff = '_{}'.format(step)\n",
    "                for f1, f2 in zip(fq, fq_base):\n",
    "                    shutil.copy2(f1, '{}.fastq'.format(os.path.join(out_dir, f2)))\n",
    "            elif len(steps_run) == len(args)-1:\n",
    "                in_suff = '_{}'.format(steps_run[-1])\n",
    "                out_suff = '_t'\n",
    "            else:\n",
    "                in_suff = '_{}'.format(steps_run[-1])\n",
    "                out_suff = '_{}'.format(step)\n",
    "            infiles = ['{}{}'.format(x, in_suff) for x in filepaths]\n",
    "            outfiles = ['{}{}'.format(x, out_suff) for x in filepaths]\n",
    "            if len(fq) == 2:\n",
    "                files = 'in1={0}.fastq in2={1}.fastq out1={2}.fastq out2={3}.fastq'.format(infiles[0], infiles[1], outfiles[0], outfiles[1])\n",
    "            elif len(fq) == 1:\n",
    "                files = 'in={0}.fastq out1={1}.fastq'.format(infiles[0], outfiles[0])\n",
    "            else:\n",
    "                print('I can\\'t understand the input fastq_list; did you pass a nested list?')\n",
    "            cmds = [bbduk, files, *args[step]]\n",
    "            final_cmds = ' '.join(cmd for cmd in cmds)\n",
    "    \n",
    "            p = subprocess.Popen(final_cmds, stdout=subprocess.PIPE, \n",
    "                                 stderr=subprocess.PIPE, shell=True)\n",
    "            out, err = p.communicate()\n",
    "            if p.returncode == 0:\n",
    "                steps_run.append(step)\n",
    "            else:\n",
    "                print('{} file: {} step returned error {}'.format(fq[0], step, err.decode('ascii')))\n",
    "                break\n",
    "        print(fq_base)\n",
    "        if keep_intermediate_files == False:\n",
    "            for f in filepaths:\n",
    "                os.remove('{}.fastq'.format(f))\n",
    "                for step in steps_run[:-1]:\n",
    "                    os.remove('{}_{}.fastq'.format(f, step))\n",
    "    return trim_dirs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting trimming run with args:\n",
      "\n",
      "adapter_trimming: ref=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\illum_seq.fasta ktrim=l k=15 mink=4 hdist=1\n",
      "quality_trimming: qtrim=rl trimq=10\n",
      "force_trimming: ftl=10 ftr=139\n",
      "force-trim_modulo: ftm=5\n",
      "quality_filtering: maq=10 minlen=20\n",
      "kmer_filtering: ref=C:/Users/DMacKellar/Documents/Python/BioPython/BBMap/resources/phix174_ill.ref.fa k=31 hdist=1\n",
      "\n",
      "\n",
      "Read files processed:\n",
      "['ERR030872_1', 'ERR030872_2']\n"
     ]
    }
   ],
   "source": [
    "os.chdir(paths['split3'])\n",
    "trim_dirs = run_bbduk_pipeline([thyroid_paired], trim_dirs=trim_dirs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC run\n",
      "MultiQC run\n"
     ]
    }
   ],
   "source": [
    "latest_dir = trim_dirs[str(max(list(int(x) for x in trim_dirs.keys())))]\n",
    "run_fastqc(latest_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, that seems like an extremely complicated function to run the bbduk steps in a modifiable way, but I've not got a pipeline that should work.  Let's apply it to all of the BodyMap read files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fastqs(input_dir):\n",
    "    singlets = []\n",
    "    pairs = []\n",
    "    for file in os.listdir(input_dir):\n",
    "        base = os.path.splitext(file)[0].rstrip('_t')\n",
    "        ext = os.path.splitext(file)[1]\n",
    "        if ext in ['.fq', '.fastq']:\n",
    "            if base[-2:] in ['_1', '_2']:\n",
    "                pairs.append(file)\n",
    "            else:\n",
    "                filepath = os.path.join(input_dir, file)\n",
    "                singlets.append([filepath])\n",
    "    pairs2 = sorted(pairs)\n",
    "    paired = []\n",
    "    for x, y in zip(pairs2[0::2], pairs2[1::2]):\n",
    "        base_x = os.path.splitext(x)[0].rstrip('_t')[:-2]\n",
    "        base_y = os.path.splitext(y)[0].rstrip('_t')[:-2]\n",
    "        if base_x == base_y:\n",
    "            filepath_x = os.path.join(input_dir, x)\n",
    "            filepath_y = os.path.join(input_dir, y)\n",
    "            paired.append([filepath_x, filepath_y])\n",
    "\n",
    "    return singlets + paired\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting trimming run with args:\n",
      "\n",
      "adapter_trimming: ref=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\illum_seq.fasta ktrim=l k=15 mink=4 hdist=1\n",
      "quality_trimming: qtrim=rl trimq=10\n",
      "force_trimming: ftl=10 ftr=139\n",
      "force-trim_modulo: ftm=5\n",
      "quality_filtering: maq=10 minlen=20\n",
      "kmer_filtering: ref=C:/Users/DMacKellar/Documents/Python/BioPython/BBMap/resources/phix174_ill.ref.fa k=31 hdist=1\n",
      "\n",
      "\n",
      "Read files processed:\n",
      "['ERR030856']\n",
      "['ERR030857']\n",
      "['ERR030858']\n",
      "['ERR030859']\n",
      "['ERR030860']\n",
      "['ERR030861']\n",
      "['ERR030862']\n",
      "['ERR030863']\n",
      "['ERR030864']\n",
      "['ERR030865']\n",
      "['ERR030866']\n",
      "['ERR030867']\n",
      "['ERR030868']\n",
      "['ERR030869']\n",
      "['ERR030870']\n",
      "['ERR030871']\n",
      "['ERR030888']\n",
      "['ERR030889']\n",
      "['ERR030890']\n",
      "['ERR030891']\n",
      "['ERR030892']\n",
      "['ERR030893']\n",
      "['ERR030894']\n",
      "['ERR030895']\n",
      "['ERR030896']\n",
      "['ERR030897']\n",
      "['ERR030898']\n",
      "['ERR030899']\n",
      "['ERR030900']\n",
      "['ERR030901']\n",
      "['ERR030902']\n",
      "['ERR030903']\n",
      "['ERR030872_1', 'ERR030872_2']\n",
      "['ERR030873_1', 'ERR030873_2']\n",
      "['ERR030874_1', 'ERR030874_2']\n",
      "['ERR030875_1', 'ERR030875_2']\n",
      "['ERR030876_1', 'ERR030876_2']\n",
      "['ERR030877_1', 'ERR030877_2']\n",
      "['ERR030878_1', 'ERR030878_2']\n",
      "['ERR030879_1', 'ERR030879_2']\n",
      "['ERR030880_1', 'ERR030880_2']\n",
      "['ERR030881_1', 'ERR030881_2']\n",
      "['ERR030882_1', 'ERR030882_2']\n",
      "['ERR030883_1', 'ERR030883_2']\n",
      "['ERR030884_1', 'ERR030884_2']\n",
      "['ERR030885_1', 'ERR030885_2']\n",
      "['ERR030886_1', 'ERR030886_2']\n",
      "['ERR030887_1', 'ERR030887_2']\n"
     ]
    }
   ],
   "source": [
    "os.chdir(paths['split3'])\n",
    "bmap_inputs = get_fastqs(paths['split3'])\n",
    "trim_dirs = run_bbduk_pipeline(bmap_inputs, trim_dirs=trim_dirs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC run\n",
      "MultiQC run\n"
     ]
    }
   ],
   "source": [
    "latest_dir = trim_dirs[str(max(list(int(x) for x in trim_dirs.keys())))]\n",
    "run_fastqc(latest_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, that looks good.  After lots of wrangling, I'd say I'm ready to align.\n",
    "\n",
    "\n",
    "## Aligning Reads\n",
    "\n",
    "The [classic approach](https://usegalaxy.org/u/jeremy/p/galaxy-rna-seq-analysis-exercise) is to use TopHat and Cufflinks.  It [sounds like](https://www.biostars.org/p/266161/) some more recent choices are STAR and [BBMap](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmap-guide/).  Since I've already used BBDuk for trimming, I'll try the latter first.\n",
    "\n",
    "One major question is whether to align against the human genome or some previously-compiled transcriptome.  It [sounds like](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8) the former approach will catch more of the reads and have lower duplication rates (i.e., places where a read may map multiple times).  It looks like, as of 20180629, the latest stable NCBI RefSeq build for the human genome is [GRCh38.p12](https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.38).  More generally, **the** definitive resource for maintaining and iteratively improving human genome assembly and annotation data is the [Genome Reference Consortium](https://www.ncbi.nlm.nih.gov/grc/human/data); the 'GRC' part means 'Genome Reference Consortium'.  That consortium currently only handles maintaining the standard genomes for 4 organisms, so the first letter after GRC indicates the organism: humans ('h'), zebrafish ('z'), mouse ('m'), and chicken ('g' for 'Gallus').  '38' means that (as of Nov. 2017) this is the 38th such release, and the '.p12' subscript means that this release has been 'patched' a dozen times, representing more minor revisions to the sequence since that version was released.  \n",
    "\n",
    "Each such release contains multiple datasets; transcriptomes and genomes, both including sequence from the Human Genome Sequencing project that have successfully been assembled into super-assemblies that represent whole chromosomes, as well as a genome that includes additional large scaffolds that haven't been successfully assigned to a particular chromosome.  The summary data table shows that there are two main kinds of assembly within: the more succinct and the more complete.\n",
    "\n",
    "The records are hosted in the Genbank database, and therefore outside the reach of the `sra_download` function that I defined earlier.  Instead, you could use NCBI's [Batch Entrez](https://www.ncbi.nlm.nih.gov/sites/batchentrez) website, but the programmatic solution is probably to use [BioPython's EFetch](http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc126).  Note that this generally requires an API key affiliated with an NCBI account.  Furthermore, while I thought that it would be best to use the full-featured genbank files for the reference genome (to retain annotation info), the [bbmap docs](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmap-guide/) state that they only accept fasta or fastq as inputs.\n",
    "\n",
    "Finally, when it comes to setting up a BBMap run, the [seqanswers thread](http://seqanswers.com/forums/showpost.php?p=134546&postcount=9) has some useful advice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(595, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Sequence-Name</th>\n",
       "      <th>Sequence-Role</th>\n",
       "      <th>Assigned-Molecule</th>\n",
       "      <th>Assigned-Molecule-Location/Type</th>\n",
       "      <th>GenBank-Accn</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>RefSeq-Accn</th>\n",
       "      <th>Assembly-Unit</th>\n",
       "      <th>Sequence-Length</th>\n",
       "      <th>UCSC-style-name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000663.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>248956422</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>2</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000664.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>242193529</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>3</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000665.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000003.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>198295559</td>\n",
       "      <td>chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>4</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000666.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>190214555</td>\n",
       "      <td>chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000667.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>181538259</td>\n",
       "      <td>chr5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000668.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>170805979</td>\n",
       "      <td>chr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>7</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>159345973</td>\n",
       "      <td>chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>8</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000670.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>145138636</td>\n",
       "      <td>chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>9</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000671.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>138394717</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>10</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000672.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>133797422</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>11</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000673.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>135086622</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>12</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>133275309</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>13</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000675.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>114364328</td>\n",
       "      <td>chr13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>14</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000676.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000014.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>107043718</td>\n",
       "      <td>chr14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>15</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000677.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>101991189</td>\n",
       "      <td>chr15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>16</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000678.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>90338345</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>17</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000679.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>83257441</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>18</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000680.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000018.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>80373285</td>\n",
       "      <td>chr18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000681.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>58617616</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>20</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000682.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>64444167</td>\n",
       "      <td>chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>21</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000683.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>46709983</td>\n",
       "      <td>chr21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>22</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000684.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>50818468</td>\n",
       "      <td>chr22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>X</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>X</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000685.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Y</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>Y</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000686.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HSCHR1_CTG1_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270706.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187361.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>175055</td>\n",
       "      <td>chr1_KI270706v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HSCHR1_CTG2_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187362.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>32032</td>\n",
       "      <td>chr1_KI270707v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HSCHR1_CTG3_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270708.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187363.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>127682</td>\n",
       "      <td>chr1_KI270708v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HSCHR1_CTG4_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270709.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187364.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>66860</td>\n",
       "      <td>chr1_KI270709v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HSCHR1_CTG5_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270710.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187365.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>40176</td>\n",
       "      <td>chr1_KI270710v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HSCHR1_CTG6_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187366.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>42210</td>\n",
       "      <td>chr1_KI270711v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>HSCHR19LRC_PGF2_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL949753.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NW_003571061.2</td>\n",
       "      <td>ALT_REF_LOCI_8</td>\n",
       "      <td>796479</td>\n",
       "      <td>chr19_GL949753v2_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>HSCHR19_4_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270938.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187693.1</td>\n",
       "      <td>ALT_REF_LOCI_9</td>\n",
       "      <td>1066800</td>\n",
       "      <td>chr19_KI270938v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>HSCHR19KIR_FH15_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270882.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187636.1</td>\n",
       "      <td>ALT_REF_LOCI_10</td>\n",
       "      <td>248807</td>\n",
       "      <td>chr19_KI270882v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>HSCHR19KIR_G085_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270883.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187637.1</td>\n",
       "      <td>ALT_REF_LOCI_11</td>\n",
       "      <td>170399</td>\n",
       "      <td>chr19_KI270883v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>HSCHR19KIR_G085_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270884.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187638.1</td>\n",
       "      <td>ALT_REF_LOCI_12</td>\n",
       "      <td>157053</td>\n",
       "      <td>chr19_KI270884v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>HSCHR19KIR_G248_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270885.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187639.1</td>\n",
       "      <td>ALT_REF_LOCI_13</td>\n",
       "      <td>171027</td>\n",
       "      <td>chr19_KI270885v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>HSCHR19KIR_G248_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270886.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187640.1</td>\n",
       "      <td>ALT_REF_LOCI_14</td>\n",
       "      <td>204239</td>\n",
       "      <td>chr19_KI270886v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>HSCHR19KIR_GRC212_AB_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270887.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187641.1</td>\n",
       "      <td>ALT_REF_LOCI_15</td>\n",
       "      <td>209512</td>\n",
       "      <td>chr19_KI270887v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>HSCHR19KIR_GRC212_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270888.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187642.1</td>\n",
       "      <td>ALT_REF_LOCI_16</td>\n",
       "      <td>155532</td>\n",
       "      <td>chr19_KI270888v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>HSCHR19KIR_LUCE_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270889.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187643.1</td>\n",
       "      <td>ALT_REF_LOCI_17</td>\n",
       "      <td>170698</td>\n",
       "      <td>chr19_KI270889v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270890.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187644.1</td>\n",
       "      <td>ALT_REF_LOCI_18</td>\n",
       "      <td>184499</td>\n",
       "      <td>chr19_KI270890v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>HSCHR19KIR_RSH_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270891.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187645.1</td>\n",
       "      <td>ALT_REF_LOCI_19</td>\n",
       "      <td>170680</td>\n",
       "      <td>chr19_KI270891v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>HSCHR19KIR_RSH_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270914.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187668.1</td>\n",
       "      <td>ALT_REF_LOCI_20</td>\n",
       "      <td>205194</td>\n",
       "      <td>chr19_KI270914v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>HSCHR19KIR_T7526_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270915.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187669.1</td>\n",
       "      <td>ALT_REF_LOCI_21</td>\n",
       "      <td>170665</td>\n",
       "      <td>chr19_KI270915v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579</th>\n",
       "      <td>HSCHR19KIR_T7526_BDEL_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270916.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187670.1</td>\n",
       "      <td>ALT_REF_LOCI_22</td>\n",
       "      <td>184516</td>\n",
       "      <td>chr19_KI270916v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>HSCHR19KIR_ABC08_A1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270917.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187671.1</td>\n",
       "      <td>ALT_REF_LOCI_23</td>\n",
       "      <td>190932</td>\n",
       "      <td>chr19_KI270917v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270918.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187672.1</td>\n",
       "      <td>ALT_REF_LOCI_24</td>\n",
       "      <td>123111</td>\n",
       "      <td>chr19_KI270918v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270919.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187673.1</td>\n",
       "      <td>ALT_REF_LOCI_25</td>\n",
       "      <td>170701</td>\n",
       "      <td>chr19_KI270919v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>HSCHR19KIR_FH05_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270920.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187674.1</td>\n",
       "      <td>ALT_REF_LOCI_26</td>\n",
       "      <td>198005</td>\n",
       "      <td>chr19_KI270920v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>HSCHR19KIR_FH05_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270921.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187675.1</td>\n",
       "      <td>ALT_REF_LOCI_27</td>\n",
       "      <td>282224</td>\n",
       "      <td>chr19_KI270921v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>HSCHR19KIR_FH06_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270922.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187676.1</td>\n",
       "      <td>ALT_REF_LOCI_28</td>\n",
       "      <td>187935</td>\n",
       "      <td>chr19_KI270922v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>HSCHR19KIR_FH06_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270923.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187677.1</td>\n",
       "      <td>ALT_REF_LOCI_29</td>\n",
       "      <td>189352</td>\n",
       "      <td>chr19_KI270923v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>HSCHR19KIR_FH08_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270929.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187683.1</td>\n",
       "      <td>ALT_REF_LOCI_30</td>\n",
       "      <td>186203</td>\n",
       "      <td>chr19_KI270929v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>HSCHR19KIR_FH08_BAX_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270930.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187684.1</td>\n",
       "      <td>ALT_REF_LOCI_31</td>\n",
       "      <td>200773</td>\n",
       "      <td>chr19_KI270930v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>589</th>\n",
       "      <td>HSCHR19KIR_FH13_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270931.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187685.1</td>\n",
       "      <td>ALT_REF_LOCI_32</td>\n",
       "      <td>170148</td>\n",
       "      <td>chr19_KI270931v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>HSCHR19KIR_FH13_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270932.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187686.1</td>\n",
       "      <td>ALT_REF_LOCI_33</td>\n",
       "      <td>215732</td>\n",
       "      <td>chr19_KI270932v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>HSCHR19KIR_FH15_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270933.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187687.1</td>\n",
       "      <td>ALT_REF_LOCI_34</td>\n",
       "      <td>170537</td>\n",
       "      <td>chr19_KI270933v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>HSCHR19KIR_RP5_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL000209.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_113949.2</td>\n",
       "      <td>ALT_REF_LOCI_35</td>\n",
       "      <td>177381</td>\n",
       "      <td>chr19_GL000209v2_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>MT</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>MT</td>\n",
       "      <td>Mitochondrion</td>\n",
       "      <td>J01415.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>non-nuclear</td>\n",
       "      <td>16569</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>HSCHRUN_RANDOM_CTG29</td>\n",
       "      <td>unplaced-scaffold</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>&lt;&gt;</td>\n",
       "      <td>na</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>27745</td>\n",
       "      <td>chrUn_KI270752v1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>595 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        # Sequence-Name         Sequence-Role  \\\n",
       "0                                     1    assembled-molecule   \n",
       "1                                     2    assembled-molecule   \n",
       "2                                     3    assembled-molecule   \n",
       "3                                     4    assembled-molecule   \n",
       "4                                     5    assembled-molecule   \n",
       "5                                     6    assembled-molecule   \n",
       "6                                     7    assembled-molecule   \n",
       "7                                     8    assembled-molecule   \n",
       "8                                     9    assembled-molecule   \n",
       "9                                    10    assembled-molecule   \n",
       "10                                   11    assembled-molecule   \n",
       "11                                   12    assembled-molecule   \n",
       "12                                   13    assembled-molecule   \n",
       "13                                   14    assembled-molecule   \n",
       "14                                   15    assembled-molecule   \n",
       "15                                   16    assembled-molecule   \n",
       "16                                   17    assembled-molecule   \n",
       "17                                   18    assembled-molecule   \n",
       "18                                   19    assembled-molecule   \n",
       "19                                   20    assembled-molecule   \n",
       "20                                   21    assembled-molecule   \n",
       "21                                   22    assembled-molecule   \n",
       "22                                    X    assembled-molecule   \n",
       "23                                    Y    assembled-molecule   \n",
       "24              HSCHR1_CTG1_UNLOCALIZED  unlocalized-scaffold   \n",
       "25              HSCHR1_CTG2_UNLOCALIZED  unlocalized-scaffold   \n",
       "26              HSCHR1_CTG3_UNLOCALIZED  unlocalized-scaffold   \n",
       "27              HSCHR1_CTG4_UNLOCALIZED  unlocalized-scaffold   \n",
       "28              HSCHR1_CTG5_UNLOCALIZED  unlocalized-scaffold   \n",
       "29              HSCHR1_CTG6_UNLOCALIZED  unlocalized-scaffold   \n",
       "..                                  ...                   ...   \n",
       "565              HSCHR19LRC_PGF2_CTG3_1          alt-scaffold   \n",
       "566                    HSCHR19_4_CTG3_1          alt-scaffold   \n",
       "567        HSCHR19KIR_FH15_B_HAP_CTG3_1          alt-scaffold   \n",
       "568        HSCHR19KIR_G085_A_HAP_CTG3_1          alt-scaffold   \n",
       "569      HSCHR19KIR_G085_BA1_HAP_CTG3_1          alt-scaffold   \n",
       "570        HSCHR19KIR_G248_A_HAP_CTG3_1          alt-scaffold   \n",
       "571      HSCHR19KIR_G248_BA2_HAP_CTG3_1          alt-scaffold   \n",
       "572     HSCHR19KIR_GRC212_AB_HAP_CTG3_1          alt-scaffold   \n",
       "573    HSCHR19KIR_GRC212_BA1_HAP_CTG3_1          alt-scaffold   \n",
       "574        HSCHR19KIR_LUCE_A_HAP_CTG3_1          alt-scaffold   \n",
       "575     HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1          alt-scaffold   \n",
       "576         HSCHR19KIR_RSH_A_HAP_CTG3_1          alt-scaffold   \n",
       "577       HSCHR19KIR_RSH_BA2_HAP_CTG3_1          alt-scaffold   \n",
       "578       HSCHR19KIR_T7526_A_HAP_CTG3_1          alt-scaffold   \n",
       "579    HSCHR19KIR_T7526_BDEL_HAP_CTG3_1          alt-scaffold   \n",
       "580      HSCHR19KIR_ABC08_A1_HAP_CTG3_1          alt-scaffold   \n",
       "581  HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1          alt-scaffold   \n",
       "582  HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1          alt-scaffold   \n",
       "583        HSCHR19KIR_FH05_A_HAP_CTG3_1          alt-scaffold   \n",
       "584        HSCHR19KIR_FH05_B_HAP_CTG3_1          alt-scaffold   \n",
       "585        HSCHR19KIR_FH06_A_HAP_CTG3_1          alt-scaffold   \n",
       "586      HSCHR19KIR_FH06_BA1_HAP_CTG3_1          alt-scaffold   \n",
       "587        HSCHR19KIR_FH08_A_HAP_CTG3_1          alt-scaffold   \n",
       "588      HSCHR19KIR_FH08_BAX_HAP_CTG3_1          alt-scaffold   \n",
       "589        HSCHR19KIR_FH13_A_HAP_CTG3_1          alt-scaffold   \n",
       "590      HSCHR19KIR_FH13_BA2_HAP_CTG3_1          alt-scaffold   \n",
       "591        HSCHR19KIR_FH15_A_HAP_CTG3_1          alt-scaffold   \n",
       "592         HSCHR19KIR_RP5_B_HAP_CTG3_1          alt-scaffold   \n",
       "593                                  MT    assembled-molecule   \n",
       "594                HSCHRUN_RANDOM_CTG29     unplaced-scaffold   \n",
       "\n",
       "    Assigned-Molecule Assigned-Molecule-Location/Type GenBank-Accn  \\\n",
       "0                   1                      Chromosome   CM000663.2   \n",
       "1                   2                      Chromosome   CM000664.2   \n",
       "2                   3                      Chromosome   CM000665.2   \n",
       "3                   4                      Chromosome   CM000666.2   \n",
       "4                   5                      Chromosome   CM000667.2   \n",
       "5                   6                      Chromosome   CM000668.2   \n",
       "6                   7                      Chromosome   CM000669.2   \n",
       "7                   8                      Chromosome   CM000670.2   \n",
       "8                   9                      Chromosome   CM000671.2   \n",
       "9                  10                      Chromosome   CM000672.2   \n",
       "10                 11                      Chromosome   CM000673.2   \n",
       "11                 12                      Chromosome   CM000674.2   \n",
       "12                 13                      Chromosome   CM000675.2   \n",
       "13                 14                      Chromosome   CM000676.2   \n",
       "14                 15                      Chromosome   CM000677.2   \n",
       "15                 16                      Chromosome   CM000678.2   \n",
       "16                 17                      Chromosome   CM000679.2   \n",
       "17                 18                      Chromosome   CM000680.2   \n",
       "18                 19                      Chromosome   CM000681.2   \n",
       "19                 20                      Chromosome   CM000682.2   \n",
       "20                 21                      Chromosome   CM000683.2   \n",
       "21                 22                      Chromosome   CM000684.2   \n",
       "22                  X                      Chromosome   CM000685.2   \n",
       "23                  Y                      Chromosome   CM000686.2   \n",
       "24                  1                      Chromosome   KI270706.1   \n",
       "25                  1                      Chromosome   KI270707.1   \n",
       "26                  1                      Chromosome   KI270708.1   \n",
       "27                  1                      Chromosome   KI270709.1   \n",
       "28                  1                      Chromosome   KI270710.1   \n",
       "29                  1                      Chromosome   KI270711.1   \n",
       "..                ...                             ...          ...   \n",
       "565                19                      Chromosome   GL949753.2   \n",
       "566                19                      Chromosome   KI270938.1   \n",
       "567                19                      Chromosome   KI270882.1   \n",
       "568                19                      Chromosome   KI270883.1   \n",
       "569                19                      Chromosome   KI270884.1   \n",
       "570                19                      Chromosome   KI270885.1   \n",
       "571                19                      Chromosome   KI270886.1   \n",
       "572                19                      Chromosome   KI270887.1   \n",
       "573                19                      Chromosome   KI270888.1   \n",
       "574                19                      Chromosome   KI270889.1   \n",
       "575                19                      Chromosome   KI270890.1   \n",
       "576                19                      Chromosome   KI270891.1   \n",
       "577                19                      Chromosome   KI270914.1   \n",
       "578                19                      Chromosome   KI270915.1   \n",
       "579                19                      Chromosome   KI270916.1   \n",
       "580                19                      Chromosome   KI270917.1   \n",
       "581                19                      Chromosome   KI270918.1   \n",
       "582                19                      Chromosome   KI270919.1   \n",
       "583                19                      Chromosome   KI270920.1   \n",
       "584                19                      Chromosome   KI270921.1   \n",
       "585                19                      Chromosome   KI270922.1   \n",
       "586                19                      Chromosome   KI270923.1   \n",
       "587                19                      Chromosome   KI270929.1   \n",
       "588                19                      Chromosome   KI270930.1   \n",
       "589                19                      Chromosome   KI270931.1   \n",
       "590                19                      Chromosome   KI270932.1   \n",
       "591                19                      Chromosome   KI270933.1   \n",
       "592                19                      Chromosome   GL000209.2   \n",
       "593                MT                   Mitochondrion     J01415.2   \n",
       "594                na                              na   KI270752.1   \n",
       "\n",
       "    Relationship     RefSeq-Accn     Assembly-Unit  Sequence-Length  \\\n",
       "0              =    NC_000001.11  Primary Assembly        248956422   \n",
       "1              =    NC_000002.12  Primary Assembly        242193529   \n",
       "2              =    NC_000003.12  Primary Assembly        198295559   \n",
       "3              =    NC_000004.12  Primary Assembly        190214555   \n",
       "4              =    NC_000005.10  Primary Assembly        181538259   \n",
       "5              =    NC_000006.12  Primary Assembly        170805979   \n",
       "6              =    NC_000007.14  Primary Assembly        159345973   \n",
       "7              =    NC_000008.11  Primary Assembly        145138636   \n",
       "8              =    NC_000009.12  Primary Assembly        138394717   \n",
       "9              =    NC_000010.11  Primary Assembly        133797422   \n",
       "10             =    NC_000011.10  Primary Assembly        135086622   \n",
       "11             =    NC_000012.12  Primary Assembly        133275309   \n",
       "12             =    NC_000013.11  Primary Assembly        114364328   \n",
       "13             =     NC_000014.9  Primary Assembly        107043718   \n",
       "14             =    NC_000015.10  Primary Assembly        101991189   \n",
       "15             =    NC_000016.10  Primary Assembly         90338345   \n",
       "16             =    NC_000017.11  Primary Assembly         83257441   \n",
       "17             =    NC_000018.10  Primary Assembly         80373285   \n",
       "18             =    NC_000019.10  Primary Assembly         58617616   \n",
       "19             =    NC_000020.11  Primary Assembly         64444167   \n",
       "20             =     NC_000021.9  Primary Assembly         46709983   \n",
       "21             =    NC_000022.11  Primary Assembly         50818468   \n",
       "22             =    NC_000023.11  Primary Assembly        156040895   \n",
       "23             =    NC_000024.10  Primary Assembly         57227415   \n",
       "24             =     NT_187361.1  Primary Assembly           175055   \n",
       "25             =     NT_187362.1  Primary Assembly            32032   \n",
       "26             =     NT_187363.1  Primary Assembly           127682   \n",
       "27             =     NT_187364.1  Primary Assembly            66860   \n",
       "28             =     NT_187365.1  Primary Assembly            40176   \n",
       "29             =     NT_187366.1  Primary Assembly            42210   \n",
       "..           ...             ...               ...              ...   \n",
       "565            =  NW_003571061.2    ALT_REF_LOCI_8           796479   \n",
       "566            =     NT_187693.1    ALT_REF_LOCI_9          1066800   \n",
       "567            =     NT_187636.1   ALT_REF_LOCI_10           248807   \n",
       "568            =     NT_187637.1   ALT_REF_LOCI_11           170399   \n",
       "569            =     NT_187638.1   ALT_REF_LOCI_12           157053   \n",
       "570            =     NT_187639.1   ALT_REF_LOCI_13           171027   \n",
       "571            =     NT_187640.1   ALT_REF_LOCI_14           204239   \n",
       "572            =     NT_187641.1   ALT_REF_LOCI_15           209512   \n",
       "573            =     NT_187642.1   ALT_REF_LOCI_16           155532   \n",
       "574            =     NT_187643.1   ALT_REF_LOCI_17           170698   \n",
       "575            =     NT_187644.1   ALT_REF_LOCI_18           184499   \n",
       "576            =     NT_187645.1   ALT_REF_LOCI_19           170680   \n",
       "577            =     NT_187668.1   ALT_REF_LOCI_20           205194   \n",
       "578            =     NT_187669.1   ALT_REF_LOCI_21           170665   \n",
       "579            =     NT_187670.1   ALT_REF_LOCI_22           184516   \n",
       "580            =     NT_187671.1   ALT_REF_LOCI_23           190932   \n",
       "581            =     NT_187672.1   ALT_REF_LOCI_24           123111   \n",
       "582            =     NT_187673.1   ALT_REF_LOCI_25           170701   \n",
       "583            =     NT_187674.1   ALT_REF_LOCI_26           198005   \n",
       "584            =     NT_187675.1   ALT_REF_LOCI_27           282224   \n",
       "585            =     NT_187676.1   ALT_REF_LOCI_28           187935   \n",
       "586            =     NT_187677.1   ALT_REF_LOCI_29           189352   \n",
       "587            =     NT_187683.1   ALT_REF_LOCI_30           186203   \n",
       "588            =     NT_187684.1   ALT_REF_LOCI_31           200773   \n",
       "589            =     NT_187685.1   ALT_REF_LOCI_32           170148   \n",
       "590            =     NT_187686.1   ALT_REF_LOCI_33           215732   \n",
       "591            =     NT_187687.1   ALT_REF_LOCI_34           170537   \n",
       "592            =     NT_113949.2   ALT_REF_LOCI_35           177381   \n",
       "593            =     NC_012920.1       non-nuclear            16569   \n",
       "594           <>              na  Primary Assembly            27745   \n",
       "\n",
       "            UCSC-style-name  \n",
       "0                      chr1  \n",
       "1                      chr2  \n",
       "2                      chr3  \n",
       "3                      chr4  \n",
       "4                      chr5  \n",
       "5                      chr6  \n",
       "6                      chr7  \n",
       "7                      chr8  \n",
       "8                      chr9  \n",
       "9                     chr10  \n",
       "10                    chr11  \n",
       "11                    chr12  \n",
       "12                    chr13  \n",
       "13                    chr14  \n",
       "14                    chr15  \n",
       "15                    chr16  \n",
       "16                    chr17  \n",
       "17                    chr18  \n",
       "18                    chr19  \n",
       "19                    chr20  \n",
       "20                    chr21  \n",
       "21                    chr22  \n",
       "22                     chrX  \n",
       "23                     chrY  \n",
       "24   chr1_KI270706v1_random  \n",
       "25   chr1_KI270707v1_random  \n",
       "26   chr1_KI270708v1_random  \n",
       "27   chr1_KI270709v1_random  \n",
       "28   chr1_KI270710v1_random  \n",
       "29   chr1_KI270711v1_random  \n",
       "..                      ...  \n",
       "565    chr19_GL949753v2_alt  \n",
       "566    chr19_KI270938v1_alt  \n",
       "567    chr19_KI270882v1_alt  \n",
       "568    chr19_KI270883v1_alt  \n",
       "569    chr19_KI270884v1_alt  \n",
       "570    chr19_KI270885v1_alt  \n",
       "571    chr19_KI270886v1_alt  \n",
       "572    chr19_KI270887v1_alt  \n",
       "573    chr19_KI270888v1_alt  \n",
       "574    chr19_KI270889v1_alt  \n",
       "575    chr19_KI270890v1_alt  \n",
       "576    chr19_KI270891v1_alt  \n",
       "577    chr19_KI270914v1_alt  \n",
       "578    chr19_KI270915v1_alt  \n",
       "579    chr19_KI270916v1_alt  \n",
       "580    chr19_KI270917v1_alt  \n",
       "581    chr19_KI270918v1_alt  \n",
       "582    chr19_KI270919v1_alt  \n",
       "583    chr19_KI270920v1_alt  \n",
       "584    chr19_KI270921v1_alt  \n",
       "585    chr19_KI270922v1_alt  \n",
       "586    chr19_KI270923v1_alt  \n",
       "587    chr19_KI270929v1_alt  \n",
       "588    chr19_KI270930v1_alt  \n",
       "589    chr19_KI270931v1_alt  \n",
       "590    chr19_KI270932v1_alt  \n",
       "591    chr19_KI270933v1_alt  \n",
       "592    chr19_GL000209v2_alt  \n",
       "593                    chrM  \n",
       "594        chrUn_KI270752v1  \n",
       "\n",
       "[595 rows x 10 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paths['grch38'] = os.path.join(paths['data_dir'], 'GRCh38_p12_table.txt')\n",
    "with open(paths['grch38'], 'r') as f:\n",
    "    grch38 = pd.read_csv(f, header=62, sep='\\t', encoding='utf-8')\n",
    "print(grch38.shape)\n",
    "grch38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "paths['credentials_json'] = os.path.join(paths['data_dir'], 'credentials.json')\n",
    "with open(paths['credentials_json'], 'r') as f:\n",
    "    credentials = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_chr = grch38['RefSeq-Accn'].loc[:23].tolist()\n",
    "human_chr.append(grch38['RefSeq-Accn'].loc[593])\n",
    "human_all = grch38['RefSeq-Accn'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "paths['grch38_dir'] = os.path.join(paths['data_dir'], 'grch38')\n",
    "os.chdir(paths['grch38_dir'])\n",
    "Entrez.email = credentials['ncbi']['email']\n",
    "ncbi_api = credentials['ncbi']['api_key']\n",
    "\n",
    "# sra_download(human_all)\n",
    "def download_genbank(acc_list, email=Entrez.email, api_key=ncbi_api, \n",
    "                     db='nuccore', rettype='fasta', \n",
    "                     retmode='text', file_ext='.fasta'):\n",
    "    for acc in acc_list:\n",
    "        if not os.path.isfile(acc):\n",
    "            # Downloading...\n",
    "            net_handle = Entrez.efetch(db=db, id=acc, rettype=rettype, retmode=retmode)\n",
    "            out_handle = open('{}{}'.format(acc, file_ext), \"w\")\n",
    "            out_handle.write(net_handle.read())\n",
    "            out_handle.close()\n",
    "            net_handle.close()\n",
    "            print(\"{} downloaded\".format(acc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Note: cell commented out after running once;\n",
    "# # don't want to download all those data again\n",
    "\n",
    "# download_genbank(human_chr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, I'll need to concatenate these into one large Fasta file, since I believe bbmap requires a single file as reference for mapping, not a dir."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "\n",
    "# # Note: Comment out this cell after running once\n",
    "\n",
    "# paths['grch38_dir'] = os.path.join(paths['data_dir'], 'grch38')\n",
    "# filenames = []\n",
    "\n",
    "# for f in os.listdir(paths['grch38_dir']):\n",
    "#     filenames.append(os.path.join(paths['grch38_dir'], f))\n",
    "# paths['concat'] = os.path.join(paths['grch38_dir'], 'hs.fasta')\n",
    "# with open(paths['concat'], 'w') as outfile:\n",
    "#     for fname in filenames:\n",
    "#         with open(fname) as infile:\n",
    "#             for line in infile:\n",
    "#                 outfile.write(line)\n",
    "#                 outfile.write('\\n\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the resulting file is very large: around 3.5GB.  By comparison to [a calculation](https://stackoverflow.com/a/8954593/8637821) of the total data content of the human genome, this is very large.  Anyways, let's try aligning!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "java -cp C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\BBMap\\current align2.BBMap -Xmx4G in=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\trim_2018-07-08T18_48_43\\ERR030856_t.fastq out=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\bbmappings\\manual.sam ref=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\grch38\\NC_000024.10.fasta maxindel=200k ambig=random intronlen=20 xstag=us\n",
      "java -cp C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\BBMap\\current align2.BBMap -Xmx7G ref=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\grch38\\hs.fasta usemodulo\n"
     ]
    }
   ],
   "source": [
    "bbmap_path = os.path.join(paths['bbmap_dir'], 'bbmap.sh')\n",
    "bbmap_path1 = 'java -cp {}\\current align2.BBMap'.format(paths['bbmap_dir'])\n",
    "out_path = os.path.join(*[paths['split3'], 'bbmappings', 'manual'])\n",
    "in_path = os.path.join(trim_dirs['3'], 'ERR030856_t.fastq')\n",
    "ref_path = os.path.join(paths['grch38_dir'], 'hs.fasta')\n",
    "ref_path1 = os.path.join(paths['grch38_dir'], 'NC_000024.10.fasta')\n",
    "cmd = '{0} -Xmx4G in={1} out={2}.sam ref={3} maxindel=200k ambig=random intronlen=20 xstag=us'.format(bbmap_path1, in_path, out_path, ref_path1)\n",
    "print(cmd)\n",
    "cmd1 = '{0} -Xmx7G ref={1} usemodulo'.format(bbmap_path1, ref_path)\n",
    "print(cmd1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSourceProvider</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>InsertSize</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>Library_Name</th>\n",
       "      <th>MBases</th>\n",
       "      <th>MBytes</th>\n",
       "      <th>Run</th>\n",
       "      <th>...</th>\n",
       "      <th>DATASTORE_filetype</th>\n",
       "      <th>DATASTORE_provider</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>SRA_Study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100</td>\n",
       "      <td>SAMEA962337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX011226</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>HCT20170</td>\n",
       "      <td>7290</td>\n",
       "      <td>4098</td>\n",
       "      <td>ERR030856</td>\n",
       "      <td>...</td>\n",
       "      <td>sra</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2014-05-30</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2011-03-17</td>\n",
       "      <td>ERP000546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   AvgSpotLen    BioSample BioSourceProvider Experiment  InsertSize  \\\n",
       "0         100  SAMEA962337               NaN  ERX011226           0   \n",
       "\n",
       "  LibraryLayout Library_Name  MBases  MBytes        Run    ...      \\\n",
       "0        SINGLE     HCT20170    7290    4098  ERR030856    ...       \n",
       "\n",
       "  DATASTORE_filetype DATASTORE_provider           Instrument LibrarySelection  \\\n",
       "0                sra               ncbi  Illumina HiSeq 2000             cDNA   \n",
       "\n",
       "    LibrarySource    LoadDate      Organism  Platform ReleaseDate  SRA_Study  \n",
       "0  TRANSCRIPTOMIC  2014-05-30  Homo sapiens  ILLUMINA  2011-03-17  ERP000546  \n",
       "\n",
       "[1 rows x 30 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_table.head()\n",
    "sra_table.loc[sra_table['Run'] == 'ERR030856', 'Sex']\n",
    "sra_table[sra_table['Run'] == 'ERR030856']\n",
    "# sra_table.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, what I'm learning is that my laptop lacks the RAM to build the index file for the whole human genome.  Instead, I'll have to run a scaled-down version of the mapping, using single chromosomes.  For instance, the index built fine with just `NC_000024.10` (the Y chromosome).  If desireable, it might be fine to loop through the different chromosome reference files for each read, but given this is mostly a proof-of-concept notebook, I think I'll be ok with just running the reads against a single chromosome.\n",
    "\n",
    "Actually, the mapping is relatively fast, compared to the indexing operation, at least when dealing with a few thousand reads, so in such a case the more time-permissive operation would be to loop through the chromosomes first, and for each chromosome, align all of the reads to them.\n",
    "\n",
    "I was stunned at first to run an example alignment of `ERR030856` against a reference chromosome, and get the following stats out:\n",
    "\n",
    "```\n",
    "   ------------------   Results   ------------------\n",
    "\n",
    "Genome:                 1\n",
    "Key Length:             13\n",
    "Max Indel:              200000\n",
    "Minimum Score Ratio:    0.56\n",
    "Mapping Mode:           normal\n",
    "Reads Used:             6508    (486300 bases)\n",
    "\n",
    "Mapping:                4.974 seconds.\n",
    "Reads/sec:              1308.52\n",
    "kBases/sec:             97.78\n",
    "\n",
    "\n",
    "Read 1 data:            pct reads       num reads       pct bases          num bases\n",
    "\n",
    "mapped:                   4.1026%             267         3.2387%              15750\n",
    "unambiguous:              1.7210%             112         1.7674%               8595\n",
    "ambiguous:                2.3817%             155         1.4713%               7155\n",
    "low-Q discards:           0.0000%               0         0.0000%                  0\n",
    "\n",
    "perfect best site:        0.2151%              14         0.1141%                555\n",
    "semiperfect site:         0.2151%              14         0.1141%                555\n",
    "\n",
    "Match Rate:                   NA               NA        82.6490%              14009\n",
    "Error Rate:              20.2238%             253        17.3451%               2940\n",
    "Sub Rate:                19.9840%             250         9.9941%               1694\n",
    "Del Rate:                 1.5987%              20         7.0796%               1200\n",
    "Ins Rate:                 1.5987%              20         0.2714%                 46\n",
    "N Rate:                   0.0799%               1         0.0059%                  1\n",
    "Splice Rate:              0.3197%               4       (splices at least 20 bp)\n",
    "\n",
    "Total time:             10.177 seconds.\n",
    "```\n",
    "\n",
    "4% of reads mapping successfully?  But then I realized that, of course you wouldn't expect that many reads to align to a single one of the 23 possible chromosomes.  Checking the sra_table, I find that those reads were one of the '16-tissue mixture's.\n",
    "\n",
    "That brings up a good test of accuracy: check the ovaries and testes tissues mappings to the Y chromosome (and maybe one more, for comparison, like thyroid tissue)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "java -cp C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\BBMap\\current align2.BBMap -Xmx4G in=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\trim_2018-07-08T18_48_43\\ERR030903_t.fastq out=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\bbmappings\\manual.sam ref=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\grch38\\NC_000024.10.fasta maxindel=200k ambig=random intronlen=20 xstag=us\n",
      "java -cp C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\BBMap\\current align2.BBMap -Xmx4G in=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\trim_2018-07-08T18_48_43\\ERR030902_t.fastq out=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\bbmappings\\manual.sam ref=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\grch38\\NC_000024.10.fasta maxindel=200k ambig=random intronlen=20 xstag=us\n",
      "java -cp C:\\Users\\DMacKellar\\Documents\\Python\\BioPython\\BBMap\\current align2.BBMap -Xmx4G in=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\trim_2018-07-08T18_48_43\\ERR030901_t.fastq out=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\bbmappings\\manual.sam ref=C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\grch38\\NC_000024.10.fasta maxindel=200k ambig=random intronlen=20 xstag=us\n"
     ]
    }
   ],
   "source": [
    "bbmap_path = os.path.join(paths['bbmap_dir'], 'bbmap.sh')\n",
    "bbmap_path1 = 'java -cp {}\\current align2.BBMap'.format(paths['bbmap_dir'])\n",
    "out_path = os.path.join(*[paths['split3'], 'bbmappings', 'manual'])\n",
    "thyroid_path = os.path.join(trim_dirs['3'], 'ERR030903_t.fastq')\n",
    "testes_path = os.path.join(trim_dirs['3'], 'ERR030902_t.fastq')\n",
    "ovaries_path = os.path.join(trim_dirs['3'], 'ERR030901_t.fastq')\n",
    "ref_path = os.path.join(paths['grch38_dir'], 'hs.fasta')\n",
    "ref_path1 = os.path.join(paths['grch38_dir'], 'NC_000024.10.fasta')\n",
    "cmd1 = '{0} -Xmx4G in={1} out={2}.sam ref={3} maxindel=200k ambig=random intronlen=20 xstag=us'.format(bbmap_path1, thyroid_path, out_path, ref_path1)\n",
    "cmd2 = '{0} -Xmx4G in={1} out={2}.sam ref={3} maxindel=200k ambig=random intronlen=20 xstag=us'.format(bbmap_path1, testes_path, out_path, ref_path1)\n",
    "cmd3 = '{0} -Xmx4G in={1} out={2}.sam ref={3} maxindel=200k ambig=random intronlen=20 xstag=us'.format(bbmap_path1, ovaries_path, out_path, ref_path1)\n",
    "print('{}\\n{}\\n{}'.format(cmd1, cmd2, cmd3))\n",
    "# cmd1 = '{0} -Xmx4G ref={1}'.format(bbmap_path1, ref_path1)\n",
    "# print(cmd1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For Thyroid, that gives:\n",
    "\n",
    "`\n",
    "Read 1 data:            pct reads       num reads       pct bases          num bases\n",
    "\n",
    "mapped:                   7.1513%             362         5.2927%              12225\n",
    "`\n",
    "\n",
    "For Testes, it's:\n",
    "\n",
    "`\n",
    "Read 1 data:            pct reads       num reads       pct bases          num bases\n",
    "\n",
    "mapped:                   5.4195%             290         4.7333%              12620\n",
    "`\n",
    "\n",
    "And for Ovaries:\n",
    "\n",
    "`\n",
    "Read 1 data:            pct reads       num reads       pct bases          num bases\n",
    "\n",
    "mapped:                   8.0251%             384         6.1409%              13495\n",
    "`\n",
    "\n",
    "Well, obviously that's surprising, and potentially problematic.  I'll have to check some references and see if there's any reason to expect that ovaries tissue RNA might map to the Y-chromosome.\n",
    "\n",
    "Actually, another problem occurs with the idea of mapping to different chromosomes individually: reads might align to multiple sites, and it would be hard to catch.  If this happens when using the entire genome as the index, then at least you'll be aware of which reads do so, and rank the different sites as better or worse matches.  If you map the reads to individual chromosomes serially, then it might be hard to track which ones have already mapped elsewhere with a better score.\n",
    "\n",
    "So probably the best approach to do here would be just to choose a single chromosome, map all the reads to that locally (to do some proof-of-principle in this notebook), then do a more comprehensive run on Galaxy, or some other server instance where I'll have more resources.  Let's just do Chr1.\n",
    "\n",
    "...Actually, even Chr1 runs out of memory with `-Xmx7G`.  Let's try Chr20."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Sequence-Name</th>\n",
       "      <th>Sequence-Role</th>\n",
       "      <th>Assigned-Molecule</th>\n",
       "      <th>Assigned-Molecule-Location/Type</th>\n",
       "      <th>GenBank-Accn</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>RefSeq-Accn</th>\n",
       "      <th>Assembly-Unit</th>\n",
       "      <th>Sequence-Length</th>\n",
       "      <th>UCSC-style-name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000663.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>248956422</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>2</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000664.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>242193529</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>3</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000665.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000003.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>198295559</td>\n",
       "      <td>chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>4</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000666.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>190214555</td>\n",
       "      <td>chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000667.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>181538259</td>\n",
       "      <td>chr5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000668.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>170805979</td>\n",
       "      <td>chr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>7</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>159345973</td>\n",
       "      <td>chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>8</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000670.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>145138636</td>\n",
       "      <td>chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>9</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000671.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>138394717</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>10</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000672.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>133797422</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>11</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000673.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>135086622</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>12</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>133275309</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>13</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000675.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>114364328</td>\n",
       "      <td>chr13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>14</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000676.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000014.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>107043718</td>\n",
       "      <td>chr14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>15</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000677.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>101991189</td>\n",
       "      <td>chr15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>16</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000678.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>90338345</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>17</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000679.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>83257441</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>18</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000680.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000018.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>80373285</td>\n",
       "      <td>chr18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000681.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>58617616</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>20</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000682.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>64444167</td>\n",
       "      <td>chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>21</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000683.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>46709983</td>\n",
       "      <td>chr21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>22</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000684.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>50818468</td>\n",
       "      <td>chr22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>X</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>X</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000685.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Y</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>Y</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000686.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HSCHR1_CTG1_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270706.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187361.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>175055</td>\n",
       "      <td>chr1_KI270706v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HSCHR1_CTG2_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187362.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>32032</td>\n",
       "      <td>chr1_KI270707v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HSCHR1_CTG3_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270708.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187363.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>127682</td>\n",
       "      <td>chr1_KI270708v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HSCHR1_CTG4_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270709.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187364.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>66860</td>\n",
       "      <td>chr1_KI270709v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HSCHR1_CTG5_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270710.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187365.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>40176</td>\n",
       "      <td>chr1_KI270710v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HSCHR1_CTG6_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187366.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>42210</td>\n",
       "      <td>chr1_KI270711v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>HSCHR19LRC_PGF2_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL949753.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NW_003571061.2</td>\n",
       "      <td>ALT_REF_LOCI_8</td>\n",
       "      <td>796479</td>\n",
       "      <td>chr19_GL949753v2_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>HSCHR19_4_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270938.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187693.1</td>\n",
       "      <td>ALT_REF_LOCI_9</td>\n",
       "      <td>1066800</td>\n",
       "      <td>chr19_KI270938v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>HSCHR19KIR_FH15_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270882.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187636.1</td>\n",
       "      <td>ALT_REF_LOCI_10</td>\n",
       "      <td>248807</td>\n",
       "      <td>chr19_KI270882v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>HSCHR19KIR_G085_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270883.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187637.1</td>\n",
       "      <td>ALT_REF_LOCI_11</td>\n",
       "      <td>170399</td>\n",
       "      <td>chr19_KI270883v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>HSCHR19KIR_G085_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270884.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187638.1</td>\n",
       "      <td>ALT_REF_LOCI_12</td>\n",
       "      <td>157053</td>\n",
       "      <td>chr19_KI270884v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>HSCHR19KIR_G248_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270885.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187639.1</td>\n",
       "      <td>ALT_REF_LOCI_13</td>\n",
       "      <td>171027</td>\n",
       "      <td>chr19_KI270885v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>HSCHR19KIR_G248_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270886.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187640.1</td>\n",
       "      <td>ALT_REF_LOCI_14</td>\n",
       "      <td>204239</td>\n",
       "      <td>chr19_KI270886v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>HSCHR19KIR_GRC212_AB_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270887.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187641.1</td>\n",
       "      <td>ALT_REF_LOCI_15</td>\n",
       "      <td>209512</td>\n",
       "      <td>chr19_KI270887v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>HSCHR19KIR_GRC212_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270888.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187642.1</td>\n",
       "      <td>ALT_REF_LOCI_16</td>\n",
       "      <td>155532</td>\n",
       "      <td>chr19_KI270888v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>HSCHR19KIR_LUCE_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270889.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187643.1</td>\n",
       "      <td>ALT_REF_LOCI_17</td>\n",
       "      <td>170698</td>\n",
       "      <td>chr19_KI270889v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270890.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187644.1</td>\n",
       "      <td>ALT_REF_LOCI_18</td>\n",
       "      <td>184499</td>\n",
       "      <td>chr19_KI270890v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>HSCHR19KIR_RSH_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270891.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187645.1</td>\n",
       "      <td>ALT_REF_LOCI_19</td>\n",
       "      <td>170680</td>\n",
       "      <td>chr19_KI270891v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>HSCHR19KIR_RSH_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270914.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187668.1</td>\n",
       "      <td>ALT_REF_LOCI_20</td>\n",
       "      <td>205194</td>\n",
       "      <td>chr19_KI270914v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>HSCHR19KIR_T7526_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270915.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187669.1</td>\n",
       "      <td>ALT_REF_LOCI_21</td>\n",
       "      <td>170665</td>\n",
       "      <td>chr19_KI270915v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579</th>\n",
       "      <td>HSCHR19KIR_T7526_BDEL_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270916.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187670.1</td>\n",
       "      <td>ALT_REF_LOCI_22</td>\n",
       "      <td>184516</td>\n",
       "      <td>chr19_KI270916v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>HSCHR19KIR_ABC08_A1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270917.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187671.1</td>\n",
       "      <td>ALT_REF_LOCI_23</td>\n",
       "      <td>190932</td>\n",
       "      <td>chr19_KI270917v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270918.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187672.1</td>\n",
       "      <td>ALT_REF_LOCI_24</td>\n",
       "      <td>123111</td>\n",
       "      <td>chr19_KI270918v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270919.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187673.1</td>\n",
       "      <td>ALT_REF_LOCI_25</td>\n",
       "      <td>170701</td>\n",
       "      <td>chr19_KI270919v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>HSCHR19KIR_FH05_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270920.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187674.1</td>\n",
       "      <td>ALT_REF_LOCI_26</td>\n",
       "      <td>198005</td>\n",
       "      <td>chr19_KI270920v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>HSCHR19KIR_FH05_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270921.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187675.1</td>\n",
       "      <td>ALT_REF_LOCI_27</td>\n",
       "      <td>282224</td>\n",
       "      <td>chr19_KI270921v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>HSCHR19KIR_FH06_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270922.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187676.1</td>\n",
       "      <td>ALT_REF_LOCI_28</td>\n",
       "      <td>187935</td>\n",
       "      <td>chr19_KI270922v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>HSCHR19KIR_FH06_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270923.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187677.1</td>\n",
       "      <td>ALT_REF_LOCI_29</td>\n",
       "      <td>189352</td>\n",
       "      <td>chr19_KI270923v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>HSCHR19KIR_FH08_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270929.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187683.1</td>\n",
       "      <td>ALT_REF_LOCI_30</td>\n",
       "      <td>186203</td>\n",
       "      <td>chr19_KI270929v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>HSCHR19KIR_FH08_BAX_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270930.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187684.1</td>\n",
       "      <td>ALT_REF_LOCI_31</td>\n",
       "      <td>200773</td>\n",
       "      <td>chr19_KI270930v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>589</th>\n",
       "      <td>HSCHR19KIR_FH13_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270931.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187685.1</td>\n",
       "      <td>ALT_REF_LOCI_32</td>\n",
       "      <td>170148</td>\n",
       "      <td>chr19_KI270931v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>HSCHR19KIR_FH13_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270932.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187686.1</td>\n",
       "      <td>ALT_REF_LOCI_33</td>\n",
       "      <td>215732</td>\n",
       "      <td>chr19_KI270932v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>HSCHR19KIR_FH15_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270933.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187687.1</td>\n",
       "      <td>ALT_REF_LOCI_34</td>\n",
       "      <td>170537</td>\n",
       "      <td>chr19_KI270933v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>HSCHR19KIR_RP5_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL000209.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_113949.2</td>\n",
       "      <td>ALT_REF_LOCI_35</td>\n",
       "      <td>177381</td>\n",
       "      <td>chr19_GL000209v2_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>MT</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>MT</td>\n",
       "      <td>Mitochondrion</td>\n",
       "      <td>J01415.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>non-nuclear</td>\n",
       "      <td>16569</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>HSCHRUN_RANDOM_CTG29</td>\n",
       "      <td>unplaced-scaffold</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>&lt;&gt;</td>\n",
       "      <td>na</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>27745</td>\n",
       "      <td>chrUn_KI270752v1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>595 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        # Sequence-Name         Sequence-Role  \\\n",
       "0                                     1    assembled-molecule   \n",
       "1                                     2    assembled-molecule   \n",
       "2                                     3    assembled-molecule   \n",
       "3                                     4    assembled-molecule   \n",
       "4                                     5    assembled-molecule   \n",
       "5                                     6    assembled-molecule   \n",
       "6                                     7    assembled-molecule   \n",
       "7                                     8    assembled-molecule   \n",
       "8                                     9    assembled-molecule   \n",
       "9                                    10    assembled-molecule   \n",
       "10                                   11    assembled-molecule   \n",
       "11                                   12    assembled-molecule   \n",
       "12                                   13    assembled-molecule   \n",
       "13                                   14    assembled-molecule   \n",
       "14                                   15    assembled-molecule   \n",
       "15                                   16    assembled-molecule   \n",
       "16                                   17    assembled-molecule   \n",
       "17                                   18    assembled-molecule   \n",
       "18                                   19    assembled-molecule   \n",
       "19                                   20    assembled-molecule   \n",
       "20                                   21    assembled-molecule   \n",
       "21                                   22    assembled-molecule   \n",
       "22                                    X    assembled-molecule   \n",
       "23                                    Y    assembled-molecule   \n",
       "24              HSCHR1_CTG1_UNLOCALIZED  unlocalized-scaffold   \n",
       "25              HSCHR1_CTG2_UNLOCALIZED  unlocalized-scaffold   \n",
       "26              HSCHR1_CTG3_UNLOCALIZED  unlocalized-scaffold   \n",
       "27              HSCHR1_CTG4_UNLOCALIZED  unlocalized-scaffold   \n",
       "28              HSCHR1_CTG5_UNLOCALIZED  unlocalized-scaffold   \n",
       "29              HSCHR1_CTG6_UNLOCALIZED  unlocalized-scaffold   \n",
       "..                                  ...                   ...   \n",
       "565              HSCHR19LRC_PGF2_CTG3_1          alt-scaffold   \n",
       "566                    HSCHR19_4_CTG3_1          alt-scaffold   \n",
       "567        HSCHR19KIR_FH15_B_HAP_CTG3_1          alt-scaffold   \n",
       "568        HSCHR19KIR_G085_A_HAP_CTG3_1          alt-scaffold   \n",
       "569      HSCHR19KIR_G085_BA1_HAP_CTG3_1          alt-scaffold   \n",
       "570        HSCHR19KIR_G248_A_HAP_CTG3_1          alt-scaffold   \n",
       "571      HSCHR19KIR_G248_BA2_HAP_CTG3_1          alt-scaffold   \n",
       "572     HSCHR19KIR_GRC212_AB_HAP_CTG3_1          alt-scaffold   \n",
       "573    HSCHR19KIR_GRC212_BA1_HAP_CTG3_1          alt-scaffold   \n",
       "574        HSCHR19KIR_LUCE_A_HAP_CTG3_1          alt-scaffold   \n",
       "575     HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1          alt-scaffold   \n",
       "576         HSCHR19KIR_RSH_A_HAP_CTG3_1          alt-scaffold   \n",
       "577       HSCHR19KIR_RSH_BA2_HAP_CTG3_1          alt-scaffold   \n",
       "578       HSCHR19KIR_T7526_A_HAP_CTG3_1          alt-scaffold   \n",
       "579    HSCHR19KIR_T7526_BDEL_HAP_CTG3_1          alt-scaffold   \n",
       "580      HSCHR19KIR_ABC08_A1_HAP_CTG3_1          alt-scaffold   \n",
       "581  HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1          alt-scaffold   \n",
       "582  HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1          alt-scaffold   \n",
       "583        HSCHR19KIR_FH05_A_HAP_CTG3_1          alt-scaffold   \n",
       "584        HSCHR19KIR_FH05_B_HAP_CTG3_1          alt-scaffold   \n",
       "585        HSCHR19KIR_FH06_A_HAP_CTG3_1          alt-scaffold   \n",
       "586      HSCHR19KIR_FH06_BA1_HAP_CTG3_1          alt-scaffold   \n",
       "587        HSCHR19KIR_FH08_A_HAP_CTG3_1          alt-scaffold   \n",
       "588      HSCHR19KIR_FH08_BAX_HAP_CTG3_1          alt-scaffold   \n",
       "589        HSCHR19KIR_FH13_A_HAP_CTG3_1          alt-scaffold   \n",
       "590      HSCHR19KIR_FH13_BA2_HAP_CTG3_1          alt-scaffold   \n",
       "591        HSCHR19KIR_FH15_A_HAP_CTG3_1          alt-scaffold   \n",
       "592         HSCHR19KIR_RP5_B_HAP_CTG3_1          alt-scaffold   \n",
       "593                                  MT    assembled-molecule   \n",
       "594                HSCHRUN_RANDOM_CTG29     unplaced-scaffold   \n",
       "\n",
       "    Assigned-Molecule Assigned-Molecule-Location/Type GenBank-Accn  \\\n",
       "0                   1                      Chromosome   CM000663.2   \n",
       "1                   2                      Chromosome   CM000664.2   \n",
       "2                   3                      Chromosome   CM000665.2   \n",
       "3                   4                      Chromosome   CM000666.2   \n",
       "4                   5                      Chromosome   CM000667.2   \n",
       "5                   6                      Chromosome   CM000668.2   \n",
       "6                   7                      Chromosome   CM000669.2   \n",
       "7                   8                      Chromosome   CM000670.2   \n",
       "8                   9                      Chromosome   CM000671.2   \n",
       "9                  10                      Chromosome   CM000672.2   \n",
       "10                 11                      Chromosome   CM000673.2   \n",
       "11                 12                      Chromosome   CM000674.2   \n",
       "12                 13                      Chromosome   CM000675.2   \n",
       "13                 14                      Chromosome   CM000676.2   \n",
       "14                 15                      Chromosome   CM000677.2   \n",
       "15                 16                      Chromosome   CM000678.2   \n",
       "16                 17                      Chromosome   CM000679.2   \n",
       "17                 18                      Chromosome   CM000680.2   \n",
       "18                 19                      Chromosome   CM000681.2   \n",
       "19                 20                      Chromosome   CM000682.2   \n",
       "20                 21                      Chromosome   CM000683.2   \n",
       "21                 22                      Chromosome   CM000684.2   \n",
       "22                  X                      Chromosome   CM000685.2   \n",
       "23                  Y                      Chromosome   CM000686.2   \n",
       "24                  1                      Chromosome   KI270706.1   \n",
       "25                  1                      Chromosome   KI270707.1   \n",
       "26                  1                      Chromosome   KI270708.1   \n",
       "27                  1                      Chromosome   KI270709.1   \n",
       "28                  1                      Chromosome   KI270710.1   \n",
       "29                  1                      Chromosome   KI270711.1   \n",
       "..                ...                             ...          ...   \n",
       "565                19                      Chromosome   GL949753.2   \n",
       "566                19                      Chromosome   KI270938.1   \n",
       "567                19                      Chromosome   KI270882.1   \n",
       "568                19                      Chromosome   KI270883.1   \n",
       "569                19                      Chromosome   KI270884.1   \n",
       "570                19                      Chromosome   KI270885.1   \n",
       "571                19                      Chromosome   KI270886.1   \n",
       "572                19                      Chromosome   KI270887.1   \n",
       "573                19                      Chromosome   KI270888.1   \n",
       "574                19                      Chromosome   KI270889.1   \n",
       "575                19                      Chromosome   KI270890.1   \n",
       "576                19                      Chromosome   KI270891.1   \n",
       "577                19                      Chromosome   KI270914.1   \n",
       "578                19                      Chromosome   KI270915.1   \n",
       "579                19                      Chromosome   KI270916.1   \n",
       "580                19                      Chromosome   KI270917.1   \n",
       "581                19                      Chromosome   KI270918.1   \n",
       "582                19                      Chromosome   KI270919.1   \n",
       "583                19                      Chromosome   KI270920.1   \n",
       "584                19                      Chromosome   KI270921.1   \n",
       "585                19                      Chromosome   KI270922.1   \n",
       "586                19                      Chromosome   KI270923.1   \n",
       "587                19                      Chromosome   KI270929.1   \n",
       "588                19                      Chromosome   KI270930.1   \n",
       "589                19                      Chromosome   KI270931.1   \n",
       "590                19                      Chromosome   KI270932.1   \n",
       "591                19                      Chromosome   KI270933.1   \n",
       "592                19                      Chromosome   GL000209.2   \n",
       "593                MT                   Mitochondrion     J01415.2   \n",
       "594                na                              na   KI270752.1   \n",
       "\n",
       "    Relationship     RefSeq-Accn     Assembly-Unit  Sequence-Length  \\\n",
       "0              =    NC_000001.11  Primary Assembly        248956422   \n",
       "1              =    NC_000002.12  Primary Assembly        242193529   \n",
       "2              =    NC_000003.12  Primary Assembly        198295559   \n",
       "3              =    NC_000004.12  Primary Assembly        190214555   \n",
       "4              =    NC_000005.10  Primary Assembly        181538259   \n",
       "5              =    NC_000006.12  Primary Assembly        170805979   \n",
       "6              =    NC_000007.14  Primary Assembly        159345973   \n",
       "7              =    NC_000008.11  Primary Assembly        145138636   \n",
       "8              =    NC_000009.12  Primary Assembly        138394717   \n",
       "9              =    NC_000010.11  Primary Assembly        133797422   \n",
       "10             =    NC_000011.10  Primary Assembly        135086622   \n",
       "11             =    NC_000012.12  Primary Assembly        133275309   \n",
       "12             =    NC_000013.11  Primary Assembly        114364328   \n",
       "13             =     NC_000014.9  Primary Assembly        107043718   \n",
       "14             =    NC_000015.10  Primary Assembly        101991189   \n",
       "15             =    NC_000016.10  Primary Assembly         90338345   \n",
       "16             =    NC_000017.11  Primary Assembly         83257441   \n",
       "17             =    NC_000018.10  Primary Assembly         80373285   \n",
       "18             =    NC_000019.10  Primary Assembly         58617616   \n",
       "19             =    NC_000020.11  Primary Assembly         64444167   \n",
       "20             =     NC_000021.9  Primary Assembly         46709983   \n",
       "21             =    NC_000022.11  Primary Assembly         50818468   \n",
       "22             =    NC_000023.11  Primary Assembly        156040895   \n",
       "23             =    NC_000024.10  Primary Assembly         57227415   \n",
       "24             =     NT_187361.1  Primary Assembly           175055   \n",
       "25             =     NT_187362.1  Primary Assembly            32032   \n",
       "26             =     NT_187363.1  Primary Assembly           127682   \n",
       "27             =     NT_187364.1  Primary Assembly            66860   \n",
       "28             =     NT_187365.1  Primary Assembly            40176   \n",
       "29             =     NT_187366.1  Primary Assembly            42210   \n",
       "..           ...             ...               ...              ...   \n",
       "565            =  NW_003571061.2    ALT_REF_LOCI_8           796479   \n",
       "566            =     NT_187693.1    ALT_REF_LOCI_9          1066800   \n",
       "567            =     NT_187636.1   ALT_REF_LOCI_10           248807   \n",
       "568            =     NT_187637.1   ALT_REF_LOCI_11           170399   \n",
       "569            =     NT_187638.1   ALT_REF_LOCI_12           157053   \n",
       "570            =     NT_187639.1   ALT_REF_LOCI_13           171027   \n",
       "571            =     NT_187640.1   ALT_REF_LOCI_14           204239   \n",
       "572            =     NT_187641.1   ALT_REF_LOCI_15           209512   \n",
       "573            =     NT_187642.1   ALT_REF_LOCI_16           155532   \n",
       "574            =     NT_187643.1   ALT_REF_LOCI_17           170698   \n",
       "575            =     NT_187644.1   ALT_REF_LOCI_18           184499   \n",
       "576            =     NT_187645.1   ALT_REF_LOCI_19           170680   \n",
       "577            =     NT_187668.1   ALT_REF_LOCI_20           205194   \n",
       "578            =     NT_187669.1   ALT_REF_LOCI_21           170665   \n",
       "579            =     NT_187670.1   ALT_REF_LOCI_22           184516   \n",
       "580            =     NT_187671.1   ALT_REF_LOCI_23           190932   \n",
       "581            =     NT_187672.1   ALT_REF_LOCI_24           123111   \n",
       "582            =     NT_187673.1   ALT_REF_LOCI_25           170701   \n",
       "583            =     NT_187674.1   ALT_REF_LOCI_26           198005   \n",
       "584            =     NT_187675.1   ALT_REF_LOCI_27           282224   \n",
       "585            =     NT_187676.1   ALT_REF_LOCI_28           187935   \n",
       "586            =     NT_187677.1   ALT_REF_LOCI_29           189352   \n",
       "587            =     NT_187683.1   ALT_REF_LOCI_30           186203   \n",
       "588            =     NT_187684.1   ALT_REF_LOCI_31           200773   \n",
       "589            =     NT_187685.1   ALT_REF_LOCI_32           170148   \n",
       "590            =     NT_187686.1   ALT_REF_LOCI_33           215732   \n",
       "591            =     NT_187687.1   ALT_REF_LOCI_34           170537   \n",
       "592            =     NT_113949.2   ALT_REF_LOCI_35           177381   \n",
       "593            =     NC_012920.1       non-nuclear            16569   \n",
       "594           <>              na  Primary Assembly            27745   \n",
       "\n",
       "            UCSC-style-name  \n",
       "0                      chr1  \n",
       "1                      chr2  \n",
       "2                      chr3  \n",
       "3                      chr4  \n",
       "4                      chr5  \n",
       "5                      chr6  \n",
       "6                      chr7  \n",
       "7                      chr8  \n",
       "8                      chr9  \n",
       "9                     chr10  \n",
       "10                    chr11  \n",
       "11                    chr12  \n",
       "12                    chr13  \n",
       "13                    chr14  \n",
       "14                    chr15  \n",
       "15                    chr16  \n",
       "16                    chr17  \n",
       "17                    chr18  \n",
       "18                    chr19  \n",
       "19                    chr20  \n",
       "20                    chr21  \n",
       "21                    chr22  \n",
       "22                     chrX  \n",
       "23                     chrY  \n",
       "24   chr1_KI270706v1_random  \n",
       "25   chr1_KI270707v1_random  \n",
       "26   chr1_KI270708v1_random  \n",
       "27   chr1_KI270709v1_random  \n",
       "28   chr1_KI270710v1_random  \n",
       "29   chr1_KI270711v1_random  \n",
       "..                      ...  \n",
       "565    chr19_GL949753v2_alt  \n",
       "566    chr19_KI270938v1_alt  \n",
       "567    chr19_KI270882v1_alt  \n",
       "568    chr19_KI270883v1_alt  \n",
       "569    chr19_KI270884v1_alt  \n",
       "570    chr19_KI270885v1_alt  \n",
       "571    chr19_KI270886v1_alt  \n",
       "572    chr19_KI270887v1_alt  \n",
       "573    chr19_KI270888v1_alt  \n",
       "574    chr19_KI270889v1_alt  \n",
       "575    chr19_KI270890v1_alt  \n",
       "576    chr19_KI270891v1_alt  \n",
       "577    chr19_KI270914v1_alt  \n",
       "578    chr19_KI270915v1_alt  \n",
       "579    chr19_KI270916v1_alt  \n",
       "580    chr19_KI270917v1_alt  \n",
       "581    chr19_KI270918v1_alt  \n",
       "582    chr19_KI270919v1_alt  \n",
       "583    chr19_KI270920v1_alt  \n",
       "584    chr19_KI270921v1_alt  \n",
       "585    chr19_KI270922v1_alt  \n",
       "586    chr19_KI270923v1_alt  \n",
       "587    chr19_KI270929v1_alt  \n",
       "588    chr19_KI270930v1_alt  \n",
       "589    chr19_KI270931v1_alt  \n",
       "590    chr19_KI270932v1_alt  \n",
       "591    chr19_KI270933v1_alt  \n",
       "592    chr19_GL000209v2_alt  \n",
       "593                    chrM  \n",
       "594        chrUn_KI270752v1  \n",
       "\n",
       "[595 rows x 10 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_bbmap(reads_dir, ref_file_path, map_dirs=None, use_defaults=True, get_stats=True, args_list=None, kwargs_dict=None):\n",
    "    reads_files = get_fastqs(reads_dir)\n",
    "    for files in reads_files:\n",
    "        file_base = files[0].split(sep='_')[0]\n",
    "        if len(files) == 2:\n",
    "            file_str = 'in1={} in2={} rcs=f'.format(files[0], files[1])\n",
    "        elif len(files) == 1:\n",
    "            file_str = 'in={}'.format(files[0])\n",
    "            \n",
    "    if map_dirs==None:\n",
    "        map_dirs = {}\n",
    "        new_dir = '1'\n",
    "    else:\n",
    "        new_dir = max([int(x) for x in trim_dirs.keys()]) + 1\n",
    "    try:\n",
    "        out_dir\n",
    "    except NameError:\n",
    "        ts = datetime.datetime.now().replace(microsecond=0).timestamp()\n",
    "        read_ts = datetime.datetime.fromtimestamp(ts).isoformat().replace(':', '_')\n",
    "        out_dir = os.path.join(os.getcwd(), 'map_{}'.format(read_ts))\n",
    "        os.mkdir(out_dir)\n",
    "    else:\n",
    "        if not os.path.isdir(out_dir):\n",
    "            os.mkdir(out_dir)\n",
    "    map_dirs[str(new_dir)] = str(out_dir)\n",
    "    os.chdir(out_dir)\n",
    "            \n",
    "    if o_s == 'Windows':\n",
    "        bbmap = 'java -cp {}\\current align2.BBMap'.format(paths['bbmap_dir'])\n",
    "    elif o_s == 'Darwin':\n",
    "        bbmap = os.path.join(paths['bbmap_dir'], 'bbmap.sh')\n",
    "    \n",
    "    def return_defaults():\n",
    "        default_kwargs = {'maxindel': '200k', 'ambig': 'random', 'intronlen': '20', 'xstag': 'us'}\n",
    "        default_args = ['-Xmx7G']\n",
    "        return default_args, default_kwargs\n",
    "\n",
    "    reads = get_fastqs(reads_dir)\n",
    "    \n",
    "    ref_base = os.path.splitext(os.path.basename(ref_file_path))[0].replace('.', '_')\n",
    "    print('Aligning to {}'.format(ref_base))\n",
    "    for read in reads:\n",
    "        if use_defaults==True:\n",
    "            args, kwargs = return_defaults()\n",
    "        else:\n",
    "            args, kwargs = [], {}\n",
    "\n",
    "        if args_list:\n",
    "            # imperfect, but try to replace any arg matching first 3 chars\n",
    "            for a1 in args_list:\n",
    "                for a2 in args:\n",
    "                    if a1[:3] == a2[:3]:\n",
    "                        args.remove(a2)\n",
    "                        args.append(a1)\n",
    "        if kwargs_dict:\n",
    "            for k, v in kwargs_dict.items():\n",
    "                kwargs[k] = v\n",
    "        args.append('ref={}'.format(ref_file_path))\n",
    "        if len(read) == 2:\n",
    "            base_name = [os.path.splitext(os.path.basename(x))[0].rstrip('_t') for x in read]\n",
    "            files = 'in1={0} in2={1} out={2}_{3}_{4}.sam'.format(read[0], read[1], ref_base, base_name[0], base_name[1])\n",
    "        elif len(read) == 1:\n",
    "            base_name = [os.path.splitext(os.path.basename(*read))[0].rstrip('_t')]\n",
    "            files = 'in={0} out={1}_{2}.sam'.format(read[0], ref_base, base_name[0])\n",
    "        else:\n",
    "            print('I can\\'t understand the input fastq_list; did you pass a nested list?')\n",
    "        if get_stats==True:\n",
    "            for stat in ['covstats', 'covhist', 'bincov']:\n",
    "                kwargs['{}'.format(stat)] = '{}_{}_{}.txt'.format(stat, ref_base, base_name[0])\n",
    "            for k, v in kwargs.items():\n",
    "                args.append('{}={}'.format(k, v))\n",
    "        cmds = [bbmap, files, *args]\n",
    "        final_cmds = ' '.join(cmd for cmd in cmds)\n",
    "    \n",
    "        p = subprocess.Popen(final_cmds, stdout=subprocess.PIPE, \n",
    "                             stderr=subprocess.PIPE, shell=True)\n",
    "        out, err = p.communicate()\n",
    "        if p.returncode == 0:\n",
    "            print('{} mapped'.format(read))\n",
    "        else:\n",
    "            print('{} file returned error: {}'.format(read, err.decode('ascii')))\n",
    "            break\n",
    "\n",
    "    return map_dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning to NC_000020_11\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030856_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030857_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030858_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030859_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030860_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030861_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030862_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030863_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030864_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030865_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030866_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030867_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030868_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030869_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030870_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030871_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030888_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030889_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030890_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030891_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030892_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030893_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030894_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030895_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030896_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030897_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030898_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030899_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030900_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030901_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030902_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030903_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030872_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030872_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030873_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030873_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030874_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030874_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030875_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030875_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030876_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030876_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030877_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030877_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030878_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030878_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030879_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030879_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030880_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030880_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030881_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030881_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030882_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030882_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030883_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030883_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030884_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030884_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030885_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030885_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030886_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030886_2_t.fastq'] mapped\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030887_1_t.fastq', 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\ERR030887_2_t.fastq'] mapped\n",
      "Wall time: 16min 53s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "os.chdir(paths['split3'])\n",
    "map_dirs = run_bbmap(latest_dir, paths['NC_000020_11_fasta'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, that takes a while to run, but does complete successfully.  I'll try MultiQC on the resulting dir to see whether it can wrap any of the stats files output by `bbmap`.\n",
    "\n",
    "---\n",
    "\n",
    "## Parsing BBMap Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_multiqc(in_dir):\n",
    "    if o_s == 'Darwin':\n",
    "        multiqc = 'multiqc .'\n",
    "    if o_s == 'Windows':\n",
    "        multiqc = 'python {} .'.format(os.path.join(*[paths['multiqc_dir'], 'scripts', 'multiqc']))\n",
    "    os.chdir(in_dir)\n",
    "    p = subprocess.Popen(multiqc, stdout=subprocess.PIPE, \n",
    "                         stderr=subprocess.PIPE, shell=True)\n",
    "    out, err = p.communicate()\n",
    "    if p.returncode != 0:\n",
    "        print('\\n returned an error: \\n{}\\n'.format(\n",
    "            err.decode('ascii')))\n",
    "    else:\n",
    "        print('MultiQC run')\n",
    "        report = os.path.join(in_dir, 'multiqc_report.html')\n",
    "        webbrowser.open(report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MultiQC run\n"
     ]
    }
   ],
   "source": [
    "paths['map_dir1'] = r'C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\map_2018-07-02T20_43_53'\n",
    "\n",
    "run_multiqc(paths['map_dir1'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, MultiQC *does* recognize data in that dir, but the only report it summarizes is the coverage histogram.  That's a relatively succinct output; for `covstats_NC_000020_11_ERR030856.txt` it reads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR030856 tissue:\n",
      "0    16 Tissues mixture\n",
      "Name: organism_part, dtype: object\n",
      "\n",
      "ERR030856 read length:\n",
      "0    100\n",
      "Name: AvgSpotLen, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ID</th>\n",
       "      <th>Avg_fold</th>\n",
       "      <th>Length</th>\n",
       "      <th>Ref_GC</th>\n",
       "      <th>Covered_percent</th>\n",
       "      <th>Covered_bases</th>\n",
       "      <th>Plus_reads</th>\n",
       "      <th>Minus_reads</th>\n",
       "      <th>Read_GC</th>\n",
       "      <th>Median_fold</th>\n",
       "      <th>Std_Dev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000020.11 Homo sapiens chromosome 20, GRCh3...</td>\n",
       "      <td>0.002</td>\n",
       "      <td>64444167</td>\n",
       "      <td>0.438</td>\n",
       "      <td>0.1904</td>\n",
       "      <td>122717</td>\n",
       "      <td>235</td>\n",
       "      <td>316</td>\n",
       "      <td>0.4428</td>\n",
       "      <td>0</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 #ID  Avg_fold    Length  \\\n",
       "0  NC_000020.11 Homo sapiens chromosome 20, GRCh3...     0.002  64444167   \n",
       "\n",
       "   Ref_GC  Covered_percent  Covered_bases  Plus_reads  Minus_reads  Read_GC  \\\n",
       "0   0.438           0.1904         122717         235          316   0.4428   \n",
       "\n",
       "   Median_fold  Std_Dev  \n",
       "0            0     0.05  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paths['example_covstats'] = os.path.join(\n",
    "    paths['map_dir1'], 'covstats_NC_000020_11_ERR030856.txt'\n",
    ")\n",
    "\n",
    "with open(paths['example_covstats'], 'r') as f:\n",
    "    ex_covstats_df = pd.read_csv(f, sep='\\t')\n",
    "    \n",
    "ex_covstats_table = sra_table[sra_table['Run'] == 'ERR030856']\n",
    "ex_covstats_tiss = ex_covstats_table['organism_part']\n",
    "ex_covstats_len = ex_covstats_table['AvgSpotLen']\n",
    "print('ERR030856 tissue:\\n{}\\n\\nERR030856 read length:\\n{}'.format(\n",
    "    ex_covstats_tiss, ex_covstats_len)\n",
    "     )\n",
    "ex_covstats_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In other words, for the 100bp reads run made with a mix of RNA from 16 tissues, using just the first 10,000 reads (trimmed down to ~5,000 reads), about 551 reads mapped to the 20th chromosome of the grch38 build of the *H. sapiens* genome, covering about 122,717bp of the 64,444,167bp in that chromosome, or about 0.19%.\n",
    "\n",
    "That's the `covhist` output.  BBMap can also output a `bincov` stats file, which (according to [this wiki](https://wiki.gacrc.uga.edu/wiki/BBMap#Documentation)) instead breaks down the reference file into chunks of different bp-long locations along the chromosome, and tells how many reads mapped to that region, specifically."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    64445.000000\n",
      "mean         0.002023\n",
      "std          0.039496\n",
      "min          0.000000\n",
      "25%          0.000000\n",
      "50%          0.000000\n",
      "75%          0.000000\n",
      "max          1.330000\n",
      "Name: Cov, dtype: float64\n",
      "0.00    63943\n",
      "0.01       10\n",
      "0.02       72\n",
      "0.03       37\n",
      "0.04       26\n",
      "0.05       24\n",
      "0.06       27\n",
      "0.07       19\n",
      "0.08       45\n",
      "0.09       78\n",
      "0.10        3\n",
      "0.11        1\n",
      "0.12        2\n",
      "0.13        6\n",
      "0.14        4\n",
      "0.15        1\n",
      "0.16        1\n",
      "0.17        7\n",
      "0.18        3\n",
      "0.19        3\n",
      "0.20        1\n",
      "0.21        1\n",
      "0.22        3\n",
      "0.23        2\n",
      "0.25        3\n",
      "0.26        2\n",
      "0.27        1\n",
      "0.28        3\n",
      "0.29        1\n",
      "0.31        1\n",
      "0.32        1\n",
      "0.33        2\n",
      "0.34        2\n",
      "0.35        2\n",
      "0.39        1\n",
      "0.40        2\n",
      "0.44        2\n",
      "0.45        1\n",
      "0.49        1\n",
      "0.54        1\n",
      "0.61        1\n",
      "0.62        2\n",
      "0.63        1\n",
      "0.65        2\n",
      "0.71        2\n",
      "0.74        1\n",
      "0.76        1\n",
      "0.79        1\n",
      "0.82        2\n",
      "0.84        1\n",
      "0.87        1\n",
      "1.00       80\n",
      "1.04        1\n",
      "1.15        1\n",
      "1.22        1\n",
      "1.29        1\n",
      "1.33        1\n",
      "Name: Cov, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#RefName</th>\n",
       "      <th>Cov</th>\n",
       "      <th>Pos</th>\n",
       "      <th>RunningPos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000020.11 Homo sapiens chromosome 20, GRCh3...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000020.11 Homo sapiens chromosome 20, GRCh3...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2000</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000020.11 Homo sapiens chromosome 20, GRCh3...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3000</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000020.11 Homo sapiens chromosome 20, GRCh3...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4000</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000020.11 Homo sapiens chromosome 20, GRCh3...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5000</td>\n",
       "      <td>4000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            #RefName  Cov   Pos  RunningPos\n",
       "0  NC_000020.11 Homo sapiens chromosome 20, GRCh3...  0.0  1000           0\n",
       "1  NC_000020.11 Homo sapiens chromosome 20, GRCh3...  0.0  2000        1000\n",
       "2  NC_000020.11 Homo sapiens chromosome 20, GRCh3...  0.0  3000        2000\n",
       "3  NC_000020.11 Homo sapiens chromosome 20, GRCh3...  0.0  4000        3000\n",
       "4  NC_000020.11 Homo sapiens chromosome 20, GRCh3...  0.0  5000        4000"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paths['example_bincov'] = os.path.join(\n",
    "    paths['map_dir1'], 'bincov_NC_000020_11_ERR030856.txt'\n",
    ")\n",
    "\n",
    "with open(paths['example_bincov'], 'r') as f:\n",
    "    ex_bincov_df = pd.read_csv(f, sep='\\t', header=2)\n",
    "\n",
    "print(ex_bincov_df['Cov'].describe())\n",
    "print(ex_bincov_df['Cov'].value_counts().sort_index())\n",
    "ex_bincov_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In other words, this file shows every kb of Chr20, and tells how much coverage it saw (I'm guessing that, since most of the values are fractional, it's not saying whether *any* reads mapped within that region at all, but, rather, `Cov` is a measure of how many of the 1,000bp within that region are represented).\n",
    "\n",
    "Ok, next is `covhist`, which looks like it shows pretty much the same info, but the other way around: it shows the number of bp in the whole chromosome that are covered 0, 1, 2, etc. times:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAENCAYAAAD+CUlOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4xLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvAOZPmwAAIABJREFUeJzt3X9YVFX+B/D3wIjogvy4IKNCKbimaCkjKUImv7LyR5IlbZs9GamhqUtZrYvZN/cbLWoEmaLVElq6rqZmZrltrJoKuvIjNEETynYliB8zmhiKjHe+f/A43yaBOQNzmQHfr+fpebhn5t7zOfTEp/PjnqMyGo1GEBERtcHJ3gEQEZHjY7IgIiKLmCyIiMgiJgsiIrKIyYKIiCxisiAiIouYLIiIyCImCyIisojJgoiILGKyICIii9T2DqAldXV1yMrKgru7O/r374+4uDih+yorKxWOrJmPjw/q6uo6pS57YPu6vu7eRrbPdvr37y/0vU5LFpmZmSgqKoKHhwfS0tJM5cXFxcjOzoYsy4iJiUFcXBwqKyuh1Wpxzz33YM2aNZ0VIhERtaLThqEiIyORnJxsVibLMrKyspCcnIz09HTk5uaioqICgwYNQl5eHpYvX47hw4d3VohERNSKVnsW8+bNE3rAunXrhL4XHByMmpoas7Ly8nJoNBr4+fkBAMLDw5Gfnw9nZ2fMmDEDwcHBSEtLQ1RUlFAdRESkjFaTxcKFC00/l5eX48svv8T9998PX19f1NbW4vPPP8fdd9/docr1ej0kSTJdS5KEsrIy3HPPPfjwww9x+PBh+Pr6tnp/Tk4OcnJyAACpqanw8fHpUDyi1Gp1p9VlD2xf19fd28j2db5Wk0VwcLDp56ysLCxduhTe3t6mspCQELz22muYOnVquytv6SgNlUqFW265BYsXL7Z4f2xsLDw9PVFYWAgAnTYhxMm1rq27tw/o/m1k+2zHphPcer0erq6uZmWurq7Q6/XWR/YLkiRBp9OZrnU6Hby8vKx6RmhoKEJDQzsUBxERtU1ogjs0NBQrVqzAiRMnUFFRgePHj+P111/H6NGjO1R5UFAQqqqqUFNTA4PBgLy8PKv/8BcUFODtt9/uUBxERNQ2oZ7FnDlz8OGHH+Ldd9+FXq+Ht7c3wsLCMGPGDOGKMjIyUFpaivr6eiQmJiI+Ph7R0dFISEhASkoKZFlGVFQUAgIC2t0YIiJShqo7ncHd3pfyrs15wMaR/D/nd3cr9mylcDy46+vubWT7bEd0zkL4PYsTJ05g3bp1SE1NBQB8++23OHnyZPuisyEOQxERKU8oWezduxfvvvsu+vXrh1OnTgEAXFxc8Pe//13R4ESEhobi6aeftncYRETdmlCy+Oyzz7Bs2TLExcXByan5lgEDBnTaXkxtYc+CiEh5QhPcly9fvuEFEYPBALXa/vsQcuksEZHyhHoWw4YNw65du8zK9u7dy32biIhuEkLJIiEhAceOHcMzzzyDK1eu4A9/+AOOHj2KJ554Qun4LOIwFBGR8oTGkby8vPCXv/wF3377LWprayFJEgYPHmyav7AnDkMRESlP6K/9ypUroVKpMHjwYIwbNw5DhgyBk5MTXn/9daXjIyIiByCULEpKSqwq70wchiIiUl6bw1Bbt24F0Lzy6frP11VXV7e5fXhn4TAUEZHy2kwW13eElWXZbHdYoPl19Pj4eOUiIyIih9Fmspg/fz4AYMiQIYiNje2UgIiIyPEIrYaKjY1FQ0MDKisrceXKFbPPRowYoUhgRETkOISSxYEDB5CVlQVXV1e4uLiYylUqFdasWaNYcCIKCgpQWFjI/aGIiBQklCy2bNmC5557DiEhIUrHYzVOcBMRKU9o6awsyxg5cqTSsRARkYMSShbTpk3Djh07IMuy0vEQEZEDEhqG+vTTT3HhwgXs3r0bbm5uZp+tW7dOkcCIiMhxCCWLhQsXKh2HmVOnTuHQoUOQZRkVFRV49dVXO7V+IiIyJ5QsgoODO1xRZmYmioqK4OHhgbS0NFN5cXExsrOzIcsyYmJiEBcXh2HDhmHYsGE4duwYgoKCOlw3ERF1jNCcRVNTE7Zs2YIFCxaYtiU/fvw4/vGPfwhXFBkZieTkZLMyWZaRlZWF5ORkpKenIzc3FxUVFabPDx8+jIiICOE6iIhIGULJYuPGjTh37hwWLVoElUoFAAgICMA///lP4YqCg4NvmO8oLy+HRqOBn58f1Go1wsPDkZ+fDwCoq6tD79690bt3b+E6iIhIGULDUMeOHcPq1avh6upqShbe3t7Q6/Udqlyv10OSJNO1JEkoKysDAOzbtw9RUVFt3p+Tk4OcnBwAQGpq6g1Hv4qqbtddYtobkz2p1eouGbeo7t4+oPu3ke3rfELJQq1W37Bs9uLFi3B3d+9Q5Uaj8Yay68lIZJPC2NhYeHp6orCwEEBzb8TROGJMlvj4+HTJuEV19/YB3b+NbJ/t9O/fX+h7QsNQYWFhWLNmDWpqagAA58+fR1ZWFsLDw9sfIZp7Er/czVan08HLy6tDzyQiItsTSha///3v0bdvXyxevBgNDQ1YtGgRvLy8MGPGjA5VHhQUhKqqKtTU1MBgMCAvL8/qrTtCQ0O5LxQRkcKEh6FmzZqFWbNmmYafrg8XicrIyEBpaSnq6+uRmJiI+Ph4REdHIyEhASkpKZBlGVFRUQgICLDqudxIkIhIeULJoqKiAm5ubvD09ISLiws+/PBDODk5YerUqejZs6dQRUlJSS2Wa7VaaLVa8Yh/hRsJEhEpT2gY6s0330RDQwMA4P3338epU6dw5swZvPPOO4oGJ4JncBMRKU+oZ1FbW4v+/fvDaDQiPz8faWlpcHFxwYIFC5SOzyL2LIiIlCfUs+jRowcuX76M8vJySJKEPn36oEePHmhqalI6PovYsyAiUp5QzyIiIgJ//vOfcfnyZdx3330AgLNnz6Jv376KBieCPQsiIuUJJYtZs2bh+PHjcHZ2Np25rVKpTPtEERFR9yY0DAUAI0eOhEajwZkzZwA0vyNxPXHYE4ehiIiUJ9SzqKurw5tvvonvv/8eAPDBBx/g6NGjKC4uRmJiopLxWcRhKCIi5Qn1LN555x2EhIRg48aNUKub88sdd9yBEydOKBocERE5BqFkUV5ejri4ODg5/f/Xe/fubXr3wp44DEVEpDyhYSgPDw/8+OOPZrsTVlRUOMQWuhyGIiJSnlCymDp1KlasWIG4uDjIsozDhw/jo48+QlxcnNLxERGRAxBKFtHR0XBzc8O//vUvSJKEgwcP4pFHHsGYMWOUjo+IiByAxWQhyzI+/PBDTJ8+ncmBiOgmZXGC28nJCZ9//jmcnZ07Ix6rcYKbiEh5QquhJkyYgC+++ELpWNqFhx8RESlPaM6ivLwc//jHP7B7925IkmR28NHy5csVC46IiByDULKIiYlBTEyM0rEQEZGDEkoWkZGRCodBRESOTChZAMC+ffuQm5uL8+fPw8vLCxEREYiKirL6LG4Rsixj69atuHz5MgIDA5msiIjsTChZbNq0Cfn5+Zg8eTJ8fHxQV1eHTz75BJWVlZg5c6ZQRZmZmSgqKoKHhwfS0tJM5cXFxcjOzoYsy4iJiUFcXBwKCgqg1+vh5uYGSZLa1zIiIrIZodVQBw4cwMsvv4yJEydCq9Vi4sSJWLp0Kfbv3y9cUWRkJJKTk83KZFlGVlYWkpOTkZ6ejtzcXFRUVKCyshJDhgzBE088gX/+85/WtYiIiGxOKFn06tULvXr1uqGsd+/ewhUFBwfDzc3NrKy8vBwajQZ+fn5Qq9UIDw9Hfn4+vL298Zvf/KY5QCfhIzeIiEghQsNQkyZNwuuvv464uDh4e3tDp9Nh9+7dmDx5Mqqrq03f8/Pzs6pyvV5vNswkSRLKysowadIkvPfeezh9+jSGDRvW6v05OTnIyckBAKSmprZ7Y8Nqy19pN0fYbNFaarW6S8Ytqru3D+j+bWT7Op9QstiwYQMAoKSkxKz85MmTyM7ONl1v3brVqsqNRuMNZSqVCj179sS8efMs3h8bGwtPT08UFhYCaD6kydE4YkyWXJ+X6q66e/uA7t9Gts92frmbeFuEkoW1SUCUJEnQ6XSma51OBy8vL0XqIiKi9rNqQqCurs50BrctBAUFoaqqCjU1NTAYDMjLy7P6bApu90FEpLxOO4M7IyMDpaWlqK+vR2JiIuLj4xEdHY2EhASkpKRAlmVERUUhICDAqgYUFBSgsLCQCYOISEFCyeL6GdzLly/HU089BaD5DO73339fuKKkpKQWy7VaLbRarfBzfo0n5RERKY9ncBMRkUU8g5uIiCwS6llcP4N7//79pjO409PTMW3aNKXjs4g9CyIi5XX5M7jZsyAiUh7P4CYiIot4BjcREVnEM7iJiMginsFNREQWdfkzuPkGNxGR8rr8GdxcDUVEpDyeLERERBYxWRARkUVMFkREZFGXTxZ8z4KISHlCyWLPnj2msyzOnDmDefPmYcGCBTY9CKm9+J4FEZHyhJLFp59+ir59+wIAtmzZgilTpmD69Omms7mJiKh7E0oWDQ0N6N27Ny5fvozvv/8e999/P6Kjo1FZWal0fERE5ACE3rOQJAnffPMNzp07h2HDhsHJyQkNDQ1mhyEREVH3JZQsZs6ciTfeeANqtRqLFy8GABQVFWHw4MGKBFVSUoKtW7fC398fERERGD58uCL1EBGRGKFkodVqb1hxFBYWhrCwMOGKMjMzUVRUBA8PD6SlpZnKi4uLkZ2dDVmWERMTg7i4OKhUKri6uqKpqQmSJAnXQUREyhAeR6qoqMD27duRlZUFAKiursYPP/wgXFFkZCSSk5PNymRZRlZWFpKTk5Geno7c3FxUVFRg6NChSE5OxmOPPYZt27YJ10FERMoQ6lkcOXIEf/3rXzF27Fjk5ubiqaeewpUrV/C3v/0Ny5YtE6ooODgYNTU1ZmXl5eXQaDTw8/MDAISHhyM/Px8PPvggAMDNzQ1NTU2tPjMnJwc5OTkAgNTU1HafCV7drrvEtBRT9YPhitXn91Feh5+hVqsd4nx1pXT39gHdv41sX+cTShbbtm3DsmXLMHDgQBw5cgQAcOutt5revWgvvV5vNswkSRLKysrw73//G8ePH8fPP/+M++67r9X7Y2NjERsba7quq6vrUDxK6OyYbFGfj4+PQ/4ubaW7tw/o/m1k+2ynf//+Qt8TShY//fQTbr31VrMylUpldq5FexiNxhvKVCoVxo4di7Fjxwo9g1uUExEpT2jOIjAwEAcPHjQry83N7fBqKEmSoNPpTNc6nQ5eXl4deiYREdmeUM/iySefxKuvvop9+/ahsbERKSkpqKysxEsvvdShyoOCglBVVYWamhp4e3sjLy8PixYtsuoZPM+CiEh5QsliwIAByMjIQGFhIUaPHg1JkjB69Gi4uroKV5SRkYHS0lLU19cjMTER8fHxiI6ORkJCAlJSUiDLMqKiohAQEGBVAzgMRUSkPKFkAQA9e/ZEeHjzKp7q6mrU19dblSySkpJaLNdqtdBqtcLPISKizieULDIyMnD//ffjtttuw/79+/HXv/4VTk5OePLJJxEdHa10jG3iMBQRkfKEJrhPnjyJoKAgAM3blS9btgyvvfYadu3apWhwRETkGISShcFggFqthl6vx6VLlzB06FAEBATgp59+Ujo+i3j4ERGR8oSGoQYOHIiPPvoItbW1pvkFvV6PXr16KRqcCA5DEREpT6hnkZiYiP/+97+4evUqfve73wFoPjHvrrvuUjQ4EexZEBEpT6hnodFo8Ic//MGszNpdZ5XCngURkfKEl85euHAB5eXlqK+vN9umw96roYiISHlCyeLYsWN466230K9fP5w7dw4BAQE4d+4chg4dymRBRHQTEJqz2Lp1K+bPn4+VK1fC1dUVK1euxNy5czFo0CCl47OIcxZERMoTShZ1dXUYN26cWdmECRNu2FzQHkJDQ7nVBxGRwoSSRZ8+fXDhwgUAgK+vL86cOYPq6mrIsqxocERE5BiE5ixiYmJw+vRphIWFYfLkyVi+fDlUKhWmTJmidHxEROQAhJJFXFyc6ecJEyZg+PDhuHLlCvz9/RULjIiIHIfQMBQAyLKM06dP48iRI6irqxM+ik9pnOAmIlKeUM/iP//5D1atWoWmpiZ4e3tDr9ejR48eeP755zFw4ECFQ2wbX8ojIlKeULJYt24d7r33XkyZMgUqlQpGoxGffvop1q1bhxUrVigdIxER2ZnQMFRVVRUmT54MlUoFAFCpVJg0aRJ+/PFHRYMjIiLHIJQsQkJCUFBQYFZWUFCAkJAQRYICgCtXruCPf/wjCgsLFauDiIjEtDoM9dZbb5l6ErIsIyMjA4GBgZAkCTqdDt99951VcwWZmZkoKiqCh4cH0tLSTOXFxcXIzs6GLMuIiYkxrbz6+OOPb3gRkIiI7KPVZKHRaMyuAwICTD/7+/tj5MiRVlUUGRmJ++67D2vXrjWVybKMrKwsvPTSS5AkCX/6058QGhoKvV4Pf39/NDU1WVUHEREpo9VkMWPGDJtWFBwcjJqaGrOy8vJyaDQa+Pn5AQDCw8ORn5+PK1euoLGxERUVFXBxcUFISAicnIRX+RIRkY0Jb1GuBL1eD0mSTNeSJKGsrAxPPfUUAODAgQNwd3dvNVHk5OQgJycHAJCamgofH592xVHdrrvEtBRTZ9dnLbVabZPnOKru3j6g+7eR7et8dk0WvzwX47rr8yRA89BVW2JjY+Hp6WmaBK+rq7NpfLbQ2THZoj4fHx+H/F3aSndvH9D928j22Y7oC9Z2TRbXJ8uv0+l08PLysuoZfCmPiEh5rU4EpKenm37ev3+/IpUHBQWhqqoKNTU1MBgMyMvLs/oPP7f7ICJSXqs9i+PHj8NoNEKlUmHDhg2IiorqUEUZGRkoLS1FfX09EhMTER8fj+joaCQkJCAlJQWyLCMqKsps1RURETmGVpPF0KFD8dJLL6Ffv364evUq1qxZ0+L3FixYIFRRUlJSi+VarRZarVboGS3hMBQRkfJaTRbPPfccjh49irq6OqhUKtPyVkdTUFCAwsJCnpZHRKSgVpOFi4sL7r77bgCAwWCw+XsXtsKeBRGR8oRWQ8XHx6Oqqgq5ubnQ6/Xw9vZGREQE+vXrp3R8RETkAIReiy4oKMCSJUvwww8/wM3NDZWVlViyZMkNmwvaA1dDEREpT6hnsWXLFrzwwgsYMWKEqaykpATvvfee3YeAOAxFRKQ8oZ6FXq/HsGHDzMqGDh1q9kKdvbBnQUSkPKFkMXDgQHzyySdmZXv27LH7kapAc8+CK6GIiJQlNAw1e/ZsrFixAnv37jVt0dGzZ0+8+OKLSsdHREQOQChZDBgwAOnp6SgrKzOthho8eDDUartuLUVERJ1E+K+9s7Mzhg4dqmQs7cKX8oiIlNfluwZcDUVEpDweP0dERBZZTBZGoxHV1dWQZbkz4iEiIgdkMVmoVCo8//zznRELERE5KKE5i4EDB6KqqgoDBgxQOh6rcYJbzLU5Dwh/15ozwp3f3W19METU5Qgli+HDh+O1117DhAkTbjhEPDo6WpHARHGCm4hIeULJ4ptvvkHfvn1x6tSpGz6zd7IgIiLlCSWL//mf/1E6DiIicmDC71nU19fjq6++woULF/DAAw9Ar9fDaDRCkiSbB1VRUYHPPvsM9fX1uP322zFx4kSb10FEROKEkkVpaSnS0tIQGBiIb775Bg888AB+/PFH7N69G0uWLBGqKDMzE0VFRfDw8EBaWpqpvLi4GNnZ2ZBlGTExMYiLi4O/vz/mzp0LWZa5oywRkQMQeilvw4YNSEpKwtKlS+Hs7AwAGDx4ML799lvhiiIjI5GcnGxWJssysrKykJycjPT0dOTm5qKiogJA8yqnl19+GbfffrtwHUREpAyhZFFbW3vDH221Wo1r164JVxQcHAw3NzezsvLycmg0Gvj5+UGtViM8PBz5+fkAmlc5vfrqqzh06JBwHUREpAyhYSh/f38UFxdj1KhRprKvv/4at9xyS4cq1+v1ZnMekiShrKwMJSUl+Pe//w2DwYCQkJBW78/JyUFOTg4AIDU19YZlvaKsea/AWi3F1Nn1KVlne3/n9qRWq7tk3Nbo7m1k+zqfULJ4/PHHsWLFCoSEhODq1at45513UFhYiBdeeKFDlRuNxhvKVCoVhg8fjuHDh1u8PzY2Fp6enigsLAQA1NXVdSgeJXR2TN29Plvw8fHpknFbo7u3ke2znf79+wt9TyhZDBkyBKtWrcKhQ4fg6uoKHx8fvPbaax1eCXX9IKXrdDodvLy8rHoGX8ojIlKe8NJZb29vPPDAA6ivr4e7uztUKlWHKw8KCkJVVRVqamrg7e2NvLw8LFq0yKpncLsPIiLlCSWLn3/+Ge+99x6OHj0Kg8EAtVqNsLAwPPnkkzdMWrcmIyMDpaWlqK+vR2JiIuLj4xEdHY2EhASkpKRAlmVERUUhICCgQw0iIiLbE0oWmZmZcHJywooVK+Dr64va2lps27YNmZmZwudwJyUltViu1Wqh1WrFI/4VDkMRESlPaOlsSUkJFi5cCH9/f/Ts2RP+/v545plnUFpaqnR8FhUUFPDFPSIihQn1LPr374+amhr4+/ubyurq6oRn0ZXEngURkfJaTRb79u0z/TxixAikpKRg/PjxpiVdhw4dwt13390pQbaFE9xERMprNVn8+s1pjUaDsrIylJWVma7PnDmjbHQC2LNwTNYctmQNHrZEZB+tJgtuS05ERNcJTXBf19DQAL1eb/aPvXGCm4hIeUIT3CdOnMA777yD2traGz7bunWrzYOyBoehiIiUJ5Qs1q9fj4ceeggRERFwcXFROiYiq1kzR2LtpoqcJyESHIZqampCVFQUXF1d4eTkZPaPvXEYiohIeUJ/7SdPnoyPP/64xV1i7S00NJTLZomIFCY0DDV27FikpKRg165dcHd3N/tszZo1igRGRESOQyhZvPHGGxg6dCjGjRvHOQsiopuQULKoqanBihUrHGKOgoiIOp/QX//Q0FCcPHlS6VjahRPcRETKE+pZNDU1YeXKlRg2bBg8PDzMPluwYIEigYniexZERMoTShYBAQE8lIiI6CYmlCxmzJihdBxEROTAhJJFW/MVI0aMsFkwRF2FUrvqAnxjnByTULJYt26d2fXFixdhMBggSZJi71kcO3YMRUVFuHjxIu69916MHDlSkXqIiMgyoWSxdu1as2tZlrFjxw706tXLqsoyMzNRVFQEDw8PpKWlmcqLi4uRnZ0NWZYRExODuLg4jBkzBmPGjMGlS5fwwQcfMFkQEdlRu16ccHJywvTp0/Hxxx9bdV9kZCSSk5PNymRZRlZWFpKTk5Geno7c3FxUVFSYPt+5cyfuvffe9oRJREQ20u637E6cOGH1S3rBwcFwc3MzKysvL4dGo4Gfnx/UajXCw8ORn58Po9GITZs2YdSoUQgMDGxvmEREZANCw1Dz5s0zu7569SquXr2K2bNndzgAvV4PSZJM15IkoaysDHv37sXXX3+NhoYG/Pjjj5g4ceIN9+bk5CAnJwcAkJqaCh8fn3bFYO2W1dZoKabOrk/JOrt7fa3VaY9/h9ZQq9U2eY6jYvs6n1CyWLhwodl1z5490a9fP/Tu3bvDAbS0k61KpcKkSZMwadKkNu+NjY2Fp6cnCgsLAQB1dXUdjsfWOjsm1tf167RFfT4+Pg7534OtsH22079/f6HvCSWL4ODgDgXTFkmSoNPpTNc6nQ5eXl6K1UfUFVm7VNeang+X6pKINpNFZmZmmzerVKobhqisFRQUhKqqKtTU1MDb2xt5eXlYtGiR8P3c7oOISHltJgtvb+8Wy69evYovv/wSly5dsipZZGRkoLS0FPX19UhMTER8fDyio6ORkJCAlJQUyLKMqKgoq7YWKSgoQGFhIQ9AIiJSUJvJ4ne/+53ZtcFgQE5ODnbu3IlBgwbd8LklSUlJLZZrtVpotVqrnnUdexZERMoTmrOQZRkHDhzAjh074OPjg6SkJEXnMazBngVR92DNvAznZDqfxWSRm5uLbdu2oXfv3pg9ezZCQkI6Iy5h7FkQESmvzWTxwgsvQK/XY9q0aRgzZgxUKhWqq81zup+fn6IBWsKeBRGR8tpMFv/9738BAJs3b8bmzZtb/M7WrVttH5UV2LMgIlJem8nC3omAiIgcQ7v3hnIUPIObiEh5QquhHBmHoYiIlNflexZERKS8Lp8sOAxFRKS8VoehZFkWeoBKpYJKpbJZQNbiMBQRkfJaTRaPPvqo0ANcXFzwwQcf2CwgIiJyPK0mCxcXF7zxxhtt3mw0GvHiiy/aPCgiInIsrSaLqVOnwtfX1+IDpkyZYtOAiIjI8bSaLOLj44Ue8PDDD9ssmPbgdh9E1B7WHiglqrtuXCj0nsWv94O6rkePHvD09ISTk/0WVXGCm4hIeULJoq2T65ycnDB69GjMnj0bnp6eNguMiIgch1CyePrpp1FaWoqHH37YdJD49u3bcdtttyE4OBibN29GVlYWFi9erHS8RERkB0LjR9u2bcPcuXOh0WigVquh0WgwZ84c7NixAwMGDMD8+fNRWlqqdKxERGQnQsnCaDSitrbWrKyurs704p6rqyuuXbtms6Cqq6uxbt06pKWl2eyZRETUfkLDUJMmTcKf//xnREZGQpIk6PV67N+/H5MmTQIAFBUVYciQIW0+IzMzE0VFRfDw8DBLAsXFxcjOzoYsy4iJiUFcXBz8/Pwwb948JgsiO+JqIfoloWQxbdo03HrrrThy5AjOnj0LT09PzJs3D6NGjQIAjBkzBmPGjGnzGZGRkbjvvvuwdu1aU5ksy8jKysJLL70ESZLwpz/9CaGhofD39+9Ak4iIyNaEtygfNWqUKTm0R3BwMGpqaszKysvLodFoTEezhoeHIz8/n8mCiMjBCCULg8GAnTt34uDBgzh//jy8vLxw9913Y/r06VCr238khl6vhyRJpmtJklBWVob6+nps2bIF33//PT766CM8+OCDLd6fk5ODnJwcAEBqaip8fHzaFUfLb5HYRksxdXZ9StbZ3etrrc7uXp+Sdd6s9VlDrVbb5Dm2JPSXftOmTfj2228xZ84VW7aJAAALm0lEQVQc+Pr6ora2Fjt27EBDQwNmzZrV7sqNRuMNZSqVCu7u7pg7d67F+2NjY+Hp6YnCwkIAzZPujqazY2J9Xb9O1sf6rr+i0Bn69+8v9D2hZHH06FGsWrUK7u7upocPGjQIL7zwQoeShSRJ0Ol0pmudTgcvL692P4+IiJQhvHRWCUFBQaiqqkJNTQ0MBgPy8vKs3rojNDSU+0IRESlMqGcxbtw4rFixwuwN7h07diAsLEy4ooyMDJSWlqK+vh6JiYmIj49HdHQ0EhISkJKSAlmWERUVhYCAAKsawI0EiYiUJ5QsZs6ciR07diArKwvnz5+Ht7c3wsPDrdpxNikpqcVyrVYLrVYr/Jxf40aCRETKE0oWarUajzzyCB555BFTmdFoRElJCUaMGKFYcCLYsyAiUl679xY3GAz43//9X1vG0i6csyAiUl77X5JwEOxZEFFXYM32Kda+A9IZW6h0+WTBOQsiIuW1mSyu7ypr7WdERNS9tJksHn300c6Ko904DEVEpLw2k8WaNWs6K4524zAUEZHy2kwWvr6+nRUHERE5sFaXzv79738XesC2bdtsFkx7FBQU4O2337ZrDERE3V2rPYvPPvsM0dHRFveF2rt3L+Lj420emCgOQxERKa/VZNHY2IiFCxdafECPHj1sGhARETmeVpPF1q1bOzMOIiJyYCqjUvuPExFRt9HuvaFuZkuWLLF3CIpi+7q+7t5Gtq/zMVkQEZFFTBZERGSR8yuvvPKKvYPoigIDA+0dgqLYvq6vu7eR7etcnOAmIiKLOAxFREQWdfnzLDpTcXExsrOzIcsyYmJiEBcXZ++QbKaurg5r167FhQsXoFKpEBsbi0mTJtk7LEXIsowlS5bA29vbIVeddMTPP/+M9evX49y5c1CpVJg3bx6GDBli77BsZs+ePdi3bx9UKhUCAgIwf/58uLi42DusDsnMzERRURE8PDyQlpYGALh06RLS09NRW1sLX19fPPvss3Bzc7NrnOxZCJJlGVlZWUhOTkZ6ejpyc3NRUVFh77BsxtnZGY8//jjS09ORkpKCzz//vFu175c+++wzDBgwwN5hKCI7OxujRo1CRkYGVq1a1a3aqdfrsXfvXqSmpiItLQ2yLCMvL8/eYXVYZGQkkpOTzcp27dqF22+/HatXr8btt9+OXbt22Sm6/8dkIai8vBwajQZ+fn5Qq9UIDw9Hfn6+vcOyGS8vL9OEWq9evTBgwADo9Xo7R2V7Op0ORUVFiImJsXcoNtfQ0IBTp04hOjoaAKBWq/Gb3/zGzlHZlizLuHr1Kq5du4arV6/Cy8vL3iF1WHBw8A29hvz8fEyYMAEAMGHCBIf4W8NhKEF6vR6SJJmuJUlCWVmZHSNSTk1NDc6ePYvBgwfbOxSb27BhA2bOnInLly/bOxSbq6mpQZ8+fZCZmYn//Oc/CAwMxKxZs+Dq6mrv0GzC29sbU6dOxbx58+Di4oKRI0di5MiR9g5LET/99JMpEXp5eeHixYt2jog9C2EtLRpTqVR2iERZV65cQVpaGmbNmoXevXvbOxybKiwshIeHh8MtSbSVa9eu4ezZs5g4cSJWrlyJnj17OsTwha1cunQJ+fn5WLt2Ld5++21cuXIFBw8etHdYNw0mC0GSJEGn05mudTpdt+gC/5LBYEBaWhrGjx+PsWPH2jscm/vmm29QUFCAZ555BhkZGTh58iRWr15t77BsRpIkSJKE3/72twCAsLAwnD171s5R2c7XX3+Nvn37ok+fPlCr1Rg7dizOnDlj77AU4eHhgfPnzwMAzp8/jz59+tg5IiYLYUFBQaiqqkJNTQ0MBgPy8vK61TkaRqMR69evx4ABAzBlyhR7h6OI3//+91i/fj3Wrl2LpKQkjBgxAosWLbJ3WDbj6ekJSZJQWVkJoPmPq7+/v52jsh0fHx+UlZWhsbERRqMRX3/9dbeawP+l0NBQfPnllwCAL7/8EnfeeaedI+JLeVYpKirCxo0bIcsyoqKiMH36dHuHZDOnT5/Gyy+/jFtuucU0vPboo49Cq9XaOTJllJSU4JNPPul2S2e///57rF+/HgaDAX379sX8+fPtvuTSlrZt24a8vDw4Oztj4MCBSExM7PJn6mRkZKC0tBT19fXw8PBAfHw87rzzTqSnp6Ourg4+Pj547rnn7P7vkcmCiIgs4jAUERFZxGRBREQWMVkQEZFFTBZERGQRkwUREVnEZEFERBZxbyi6aR0+fBh79uzBDz/8gF69emHgwIGYPn06hg4dau/QiBwOkwXdlPbs2YNdu3Zhzpw5GDlyJNRqNYqLi5Gfn98pycJoNMJoNMLJiZ176hr4Uh7ddBoaGvD0009j/vz5GDdu3A2fNzU1YfPmzThy5AgAYNy4cXjsscfQo0cPPPvss5g5cyZGjx4NoHnzvrlz52Lp0qUIDAzEmTNn8P7776OiogK+vr6YNWsWhg8fDgB45ZVXcNttt6G0tBTfffcd0tLScOrUKezevRs6nQ59+vTBtGnTcM8995hi+fjjj/Hpp59CpVIhPj4eb7/9NlavXg2NRoOmpiZs2bIFR44cgcFgwJ133olZs2Z1+cOAyDHxf2vopnPmzBk0NTVhzJgxLX6+c+dOlJWVYeXKlVi1ahXKy8uxY8cOAEBERARyc3NN3z1+/Djc3d0RGBgIvV6P1NRUTJ8+He+99x4ef/xxpKWlmW0vffDgQcydOxfvv/8+fHx84OHhgT/+8Y/YuHEj5s+fj40bN+K7774D0Hwy4549e7Bs2TKsXr0apaWlZnFu3rwZVVVVWLVqFVavXg29Xo/t27fb+tdFBIDJgm5C9fX1cHd3h7Ozc4ufHz58GA899BA8PDzQp08fPPzwwzh06BAA4K677kJBQQEaGxtN373rrrsANCeCkJAQaLVaODk54Y477kBQUBCKiopMz46MjERAQACcnZ2hVquh1Wqh0WigUqkQHByMO+64A6dPnwYA5OXlISoqCgEBAejZsydmzJhheo7RaMS//vUvPPHEE3Bzc0OvXr0wffp0s0RGZEucs6Cbjru7O+rr63Ht2rUWE4Zer4evr6/p2tfX13RqoEajwYABA1BYWIjRo0ejoKAAK1euBNB8jvnRo0dRWFhouvfatWumYSgAZgdoAcBXX32F7du3o7KyEkajEY2NjbjlllsANG9NHRQU1OK9Fy9eRGNjo9lGiEajEbIst+t3QmQJkwXddIYMGYIePXogPz8fYWFhN3zu7e2N2tpaBAQEAGhOAt7e3qbPIyIicPjwYciyDH9/f2g0GgDNf8zHjx+PxMTEVuv+5YFZTU1NSEtLw4IFCxAaGgq1Wm1KPEDzCWm/PkPlOnd3d7i4uOCNN94wi41IKRyGoptO79698cgjjyArKwvHjh1DY2MjDAYDvvrqK2zatAkRERHYuXMnLl68iIsXL2L79u0YP3686f6IiAicOHECX3zxhWkICgDGjx+PwsJCFBcXm86KLikpMfsj/0sGgwFNTU3o06cPnJ2d8dVXX+HEiROmz8eNG4cDBw6goqICjY2NZvMRTk5OiImJwYYNG/DTTz8BaO4RFRcX2/rXRQSAPQu6SU2ZMgUeHh7YuXMn3nrrLbi6uiIwMBDTp0/HoEGD0NDQgOeffx5A84lzvzy7xMvLC0OGDEFpaSmeffZZU7mPjw9efPFFbNq0CW+++SacnJwwePBgzJkzp8UYevXqhSeffBLp6eloamrC6NGjzQ7UCgkJwf3334/ly5fDyckJDz30EA4ePAi1uvk/28ceewzbt2/H0qVLUV9fD29vb9xzzz0YNWqUEr8yuslx6SxRF1FRUYHFixfjb3/7W6uT80RK4TAUkQM7duwYDAYDLl26hM2bN2P06NFMFGQXHIYicmBffPEF1q5dCycnJwQHB2P27Nn2DoluUhyGIiIiizgMRUREFjFZEBGRRUwWRERkEZMFERFZxGRBREQWMVkQEZFF/wdMYBBiAJunGQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x240521a78d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Coverage</th>\n",
       "      <th>numBases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>64342285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>98783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    #Coverage  numBases\n",
       "0           0  64342285\n",
       "1           1     98783\n",
       "2           2      1965\n",
       "3           3       476\n",
       "4           4       326\n",
       "5           5       152\n",
       "6           6        81\n",
       "7           7        12\n",
       "8           8        61\n",
       "9           9        21\n",
       "10         10         5"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paths['example_covhist'] = os.path.join(paths['map_dir1'], 'covhist_NC_000020_11_ERR030857.txt')\n",
    "\n",
    "with open(paths['example_covhist'], 'r') as f:\n",
    "    ex_covhist_df = pd.read_csv(f, sep='\\t')\n",
    "\n",
    "plt.bar(ex_covhist_df['#Coverage'], ex_covhist_df['numBases'])\n",
    "plt.yscale('log')\n",
    "plt.ylabel('[Log] Number of bases represented')\n",
    "plt.xlabel('Coverage')\n",
    "plt.show()\n",
    "ex_covhist_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There was also an option to output an additional stats file with the arg `basecov=basecov.txt`, but it was too big to keep (like, multiple gigabytes).  I'll try producing a single one here, though, and parsing its content):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning to NC_000020_11\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\dummy\\\\ERR030856_t.fastq'] mapped\n",
      "#RefName\tPos\tCoverage\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t0\t0\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t1\t0\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t2\t0\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t3\t0\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t4\t0\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t5\t0\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t6\t0\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t7\t0\n",
      "\n",
      "NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\t8\t0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "os.chdir(trim_dirs['3'])\n",
    "if not os.path.exists('dummy'):\n",
    "    os.mkdir('dummy')\n",
    "os.chdir('dummy')\n",
    "shutil.copy2(os.path.join(trim_dirs['3'], 'ERR030856_t.fastq'), 'ERR030856_t.fastq')\n",
    "\n",
    "kwargs_dict = {'basecov': 'basecov.txt'}\n",
    "run_bbmap(os.getcwd(), paths['NC_000020_11_fasta'], kwargs_dict=kwargs_dict)\n",
    "\n",
    "with open('basecov.txt', 'r') as f:\n",
    "    head = [next(f) for x in range(10)]\n",
    "    \n",
    "os.chdir(trim_dirs['3'])\n",
    "with open('example_basecov.txt', 'w') as f:\n",
    "    for line in head:\n",
    "        f.write(line)\n",
    "\n",
    "shutil.rmtree('dummy')\n",
    "for line in head:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so that `basecov.txt` file is 4.79GB in size; much larger than the input reference fasta.  And it appears that BBMap may attempt to generate one such file for *every* alignment made.  That would consume *way* too much space to be worth it.  But, to summarize the output, it looks essentially like the name describing the file.  Specifically, for *every* base in the reference chromosome, it lists the coverage from the reads, explicitly, in human-readable format (except of course for the fact that no human could feasibly have the patience to iteratively search through such an enormous number of lines to retrieve the coverage they want).  That info should be recoverable through some series of operations from the output `.sam` file itself, which is way more compact; there's no justification for keeping those stats output files from BBMap."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, there's the sam file itself.  In looking for a viewer for that file format, I cam across a [thread](https://www.biostars.org/p/12752/) suggesting the use of software called `IGV`, from the [Broad Institute](http://software.broadinstitute.org/software/igv/AlignmentData).  Unfortunately, that program requires `bam` file format for input, which will require me to re-run bbmap:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning to NC_000020_11\n",
      "['C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\dummy\\\\ERR030856_t.fastq'] mapped\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'1': 'C:\\\\Users\\\\DMacKellar\\\\Documents\\\\Data\\\\Bio\\\\Bmap\\\\split3\\\\trim_2018-07-08T18_48_43\\\\dummy\\\\map_2018-07-08T19_11_38'}"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir(trim_dirs['3'])\n",
    "if not os.path.exists('dummy'):\n",
    "    os.mkdir('dummy')\n",
    "os.chdir('dummy')\n",
    "shutil.copy2(os.path.join(trim_dirs['3'], 'ERR030856_t.fastq'), 'ERR030856_t.fastq')\n",
    "\n",
    "kwargs_dict = {'out': 'ERR030856.bam'}\n",
    "run_bbmap(os.getcwd(), paths['NC_000020_11_fasta'], kwargs_dict=kwargs_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The outputs of the `BBMap` mapping are in SAM format, and in order to visualize them in the IGV viewer they must be converted to BAM and sorted.  The general suite of tools to do this is [Samtools](http://samtools.sourceforge.net/), which was written in C and which I had a lot of trouble compiling on Windows with various compilers.  So instead I defaulted to using the Windows 10 Ubuntu BASH support, which, when invoked, requires a slight alteration to the directory structure, in order to point to the files properly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "bash = 'bash'\n",
    "samtools = '/mnt/c/Users/DMacKellar/Documents/Python/BioPython/samtools-1.8/samtools'\n",
    "map_dir = '/mnt/c/Users/DMacKellar/Documents/Data/Bio/Bmap/split3/map2018-07-02T20_43_53'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_sam(input_dir):\n",
    "    os.chdir(input_dir)\n",
    "    sam_files = []\n",
    "    for file in os.listdir(input_dir):\n",
    "        if os.path.splitext(file)[-1] == '.sam':\n",
    "            sam_files.append(file)\n",
    "            \n",
    "    if o_s == 'Windows':\n",
    "        samtools = '/mnt/c/Users/DMacKellar/Documents/Python/BioPython/samtools-1.8/samtools'\n",
    "        sort_cmds = ['bash']\n",
    "        index_cmds = ['bash']\n",
    "        for sam in sam_files[::10]:\n",
    "            # subprocess can't handle arbitrarily long cmd line strings\n",
    "            fname = os.path.splitext(os.path.basename(sam))[0]\n",
    "#             print(fname)\n",
    "            sort_cmds.append('{0} sort {1} -o {2}.bam'.format(samtools, sam, fname))\n",
    "            index_cmds.append('{0} index {2}.bam'.format(samtools, sam, fname))\n",
    "        final_sort_cmds = ';'.join(sort_cmds)\n",
    "        final_index_cmds = ';'.join(index_cmds)\n",
    "        p1 = subprocess.Popen(final_sort_cmds, stdout=subprocess.PIPE, \n",
    "                         stderr=subprocess.PIPE, shell=True)\n",
    "        out, err = p1.communicate()\n",
    "        if p1.returncode != 0:\n",
    "            print('\\n returned an error: \\n{}\\n'.format(\n",
    "                err.decode('ascii')))\n",
    "        else:\n",
    "            print('Sam files sorted.')\n",
    "        p2 = subprocess.Popen(final_index_cmds, stdout=subprocess.PIPE, \n",
    "                         stderr=subprocess.PIPE, shell=True)\n",
    "        out, err = p2.communicate()\n",
    "        if p2.returncode != 0:\n",
    "            print('\\n returned an error: \\n{}\\n'.format(\n",
    "                err.decode('ascii')))\n",
    "        else:\n",
    "            print('Bam files indexed.')\n",
    "#         report = os.path.join(in_dir, 'multiqc_report.html')\n",
    "#         webbrowser.open(report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sam files sorted.\n",
      "Bam files indexed.\n"
     ]
    }
   ],
   "source": [
    "map_dir1 = r'C:\\Users\\DMacKellar\\Documents\\Data\\Bio\\Bmap\\split3\\map_2018-07-02T20_43_53'\n",
    "sort_sam(map_dir1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Confronting Resource Limitations\n",
    "\n",
    "That works, but the resulting view looks to be empty.  It turns out that 10,000 reads is going to output a very sparse map at the genomic level; typical RNAseq applications anticipate a lot higher coverage.  Zooming in to 75% of the max is the only way I can get to see anything in this first-pass, low-resolution run.\n",
    "\n",
    "This brings up a fundamental problem with prototyping an RNAseq analysis pipeline on local hardware; the memory is too limited to achieve realistic results that would help guide the process going forward, when you anticipate running a full-scale experiment.  To resolve this, I've cast about for sources of more computational power, and hit upon the Bioinformatics resource named Galaxy, which claims to offer communal resources to users for free.  I've started to document my experience with that server, but, as this notebook is already getting significantly long, will store them in another Jupyter notebook.\n",
    "\n",
    "**Note:** After a while of playing around with Galaxy, modifying my approach above to fit their BioBlend Python API, I'm finding that Galaxy also doesn't really offer something for nothing to the extent that I was hoping.  Their focus is primarily on standardizing and rendering more reproducible NGS analysis pipelines; they serve as a depot for workflows and a common syntax for using and distributing analytical approaches.  Their [storage capacity](https://galaxyproject.org/news/galaxy-public-servers-usage-quotas/) for a free account (on the 'Galaxy Main' server) is generous enough: around 200GB per round of analysis, but their memory allocation is [more limited](https://galaxyproject.org/main/#quotas).  They give you about 32GB of RAM to play with, which turns out to be enough to handle the preliminary steps of quality control on a single dataset from the Illumina BodyMap data (i.e., reads from a single tissue type), but it falls short when it comes to mapping to a reference transcriptome.\n",
    "\n",
    "The alternative is to use their interface and general approach to wrap an instance of a cloud-based, fee-requiring service (like AWS), and I haven't tried that yet, but it will have to be my next stop.  For now, however, both my locally-supported AND Galaxy-based approaches to finishing this RNAseq pipeline turn out to have insufficient resources to carry out the mapping and quantification of the BodyMap datasets to the human genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "154px",
    "width": "252px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
